{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b73e54d0-67f7-43c8-add4-9378948c687d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from collections import namedtuple\n",
    "import pandas as pd\n",
    "import json\n",
    "import time\n",
    "import copy\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4f7dd757-839d-4b1b-8de0-a3ade640d74e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa12b00e-df17-4676-8ef0-e9962ae1b6d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "t0 = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ca73c686-2f9e-4414-b976-de1cac277bfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7013</td>\n",
       "      <td>9049</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr  start   end strand  type\n",
       "0  1.0    335   649      +  gene\n",
       "1  1.0    538   792      +  gene\n",
       "2  1.0   1807  2169      -  gene\n",
       "3  1.0   2480  2707      +  gene\n",
       "4  1.0   7013  9049      -  gene"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feat = pd.read_table(\"/outs/features.txt\", sep=' ')\n",
    "# feat = pd.read_table(\"test/test_feat.txt\", sep=' ')\n",
    "feat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4af795cd-a3dd-4b09-88ce-c352d36d2b6e",
   "metadata": {},
   "source": [
    "Defining a namedtuple here that will be used for the features. Adding an antisense boolean attribute to differentiate between sense and antisense promotion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2e174706-553f-44f4-a020-f05ba8becb7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "Feat = namedtuple('Feat', ['chr', 'start', 'end', 'strand', 'type'], defaults = (False,))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21a6407e-e1be-44c4-8241-25b293f8d17e",
   "metadata": {},
   "source": [
    "Create a dictionary containing every position of each feature. Key is (chromosome, position, strand) and value is a list of namedtuples Feat."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d898c1f8-487a-4cc9-9929-73ab90bc7e51",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos = defaultdict(list)\n",
    "for row in feat.itertuples(index=False):\n",
    "    for i in range(row.start, row.end+1):\n",
    "        item = Feat(row.chr, row.start, row.end, row.strand, row.type)\n",
    "        pos[(row.chr, i, row.strand)].append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8e5dbd74-150c-442c-8f65-6190b3b234ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sorted_pos = sorted(pos.items(), key = lambda i: i[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "bf8c42d4-9520-405f-a04a-fa85bd98b98c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>value</th>\n",
       "      <th>chr</th>\n",
       "      <th>reads</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>156</td>\n",
       "      <td>0.481988</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5793</td>\n",
       "      <td>0.481988</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5871</td>\n",
       "      <td>0.481988</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5883</td>\n",
       "      <td>0.481988</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5889</td>\n",
       "      <td>0.481988</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pos     value  chr  reads strand\n",
       "0   156  0.481988    1      1      +\n",
       "1  5793  0.481988    1      1      +\n",
       "2  5871  0.481988    1      1      +\n",
       "3  5883  0.481988    1      1      +\n",
       "4  5889  0.481988    1      1      +"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "net = pd.read_table(\"/outs/net.txt\", sep=' ')\n",
    "# net = pd.read_table(\"test/net.txt\", sep=' ')\n",
    "net.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "018b01cd-dd39-4394-a492-350c72454ce7",
   "metadata": {},
   "source": [
    "First step is counting reads for each feature. If there is an overlap, I do a weighted split where the weight is the inverse of distance from the start site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7a02a8ff-61ad-4a12-a658-790d5acc9f58",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = defaultdict(float)\n",
    "ua = defaultdict(float)\n",
    "ua_dist = defaultdict(float)\n",
    "for n in net.itertuples():\n",
    "    # making the key\n",
    "    tup = (n.chr, n.pos, n.strand)\n",
    "    # check if the position is part of a feature\n",
    "    if tup in pos:\n",
    "        keys = pos.get(tup)\n",
    "        total = 0 # this is the denominator\n",
    "        checksum = 0\n",
    "        # on the + strand, check distance from start\n",
    "        if n.strand == \"+\":\n",
    "            # first find the denominator\n",
    "            for key in keys:\n",
    "                total += 1/(tup[1]-key.start + 1)                \n",
    "            # populate d with the weighted reads    \n",
    "            for key in keys:\n",
    "                d[key] += n.reads * (1/(tup[1]-key.start + 1)) / total\n",
    "                checksum += n.reads * (1/(tup[1]-key.start + 1)) / total\n",
    "        # on the - strand, check distance from end    \n",
    "        else:\n",
    "            # first find the denominator\n",
    "            for key in keys:\n",
    "                total += 1/(key.end - tup[1] + 1)\n",
    "            # populate d with the weighted reads    \n",
    "            for key in keys:\n",
    "                d[key] += n.reads * (1/(key.end - tup[1] + 1)) / total\n",
    "                checksum += n.reads * (1/(key.end - tup[1] + 1)) / total\n",
    "        # check that we assigned all reads somewhere\n",
    "        if n.reads != round(checksum):\n",
    "            print(n, \"failed checksum\")\n",
    "    # when position is not a part of any feature\n",
    "    else:\n",
    "        # find nearest feature\n",
    "        min_dist = None\n",
    "        near_feat = None\n",
    "        window_size = 600\n",
    "        keys = None\n",
    "        upper = None\n",
    "        lower = None\n",
    "        flag = False\n",
    "        flagged_keys = []\n",
    "        key2 = None\n",
    "        # checker\n",
    "        opp = (n.chr, n.pos, '+' if n.strand == \"-\" else \"-\")\n",
    "        if opp in pos:\n",
    "            flag = True\n",
    "            flagged_keys = pos.get(opp)\n",
    "            # for key in keys:\n",
    "            #     lower = min(lower, key.start) if lower is not None else key.start\n",
    "            #     upper = max(upper, key.end) if upper is not None else key.end\n",
    "        \n",
    "        for i in range(0, window_size):             \n",
    "            t_same = (n.chr, (n.pos-i) if n.strand == '+' else (n.pos+i), n.strand)\n",
    "            t_opp = (n.chr, (n.pos-i) if n.strand == '+' else (n.pos+i), '+' if n.strand == \"-\" else \"-\")\n",
    "            if t_same in pos:\n",
    "                keys = pos.get(t_same)\n",
    "                for key in keys:\n",
    "                    key2 = Feat(key.chr, key.start, key.end, key.strand, key.type+'_same')\n",
    "                    ua[key2] += n.reads/len(keys)\n",
    "                    ua_dist[key2] = max(ua_dist[key2], min(abs(n.pos-key.start), abs(n.pos-key.end))) if ua_dist[key2] != 0 \\\n",
    "                    else min(abs(n.pos-key.start), abs(n.pos-key.end))\n",
    "                if ua_dist[key2] > 1000:\n",
    "                    print(n, key2, t_same)\n",
    "                break\n",
    "\n",
    "            # if flag:\n",
    "            #     if t_opp[1] >= lower and t_opp[1] <= upper:\n",
    "            #         continue\n",
    "            \n",
    "            if t_opp in pos:\n",
    "                keys = pos.get(t_opp)\n",
    "                for key in keys:\n",
    "                    if key in flagged_keys:\n",
    "                        continue\n",
    "                    key2 = Feat(key.chr, key.start, key.end, key.strand, key.type+'_opp')\n",
    "                    ua[key2] += n.reads/len(keys)\n",
    "                    ua_dist[key2] = max(ua_dist[key2], min(abs(n.pos-key.start), abs(n.pos-key.end))) if ua_dist[key2] != 0 \\\n",
    "                    else min(abs(n.pos-key.start), abs(n.pos-key.end))\n",
    "                if key2:\n",
    "                    if ua_dist[key2] > 1000:\n",
    "                        print(n, key2, t_opp)\n",
    "                break\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "09f12586-1c52-4a53-a33f-353f864f6ce8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Feat(chr=4.0, start=149203, end=155640, strand='-', type='gene'),\n",
       " Feat(chr=4.0, start=149100, end=155932, strand='-', type='ORFs')]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos[(4, 155640, '-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "72179629-a222-414b-8240-74cc5a84e68e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4862, 4862)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ua), len(ua_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4b5eaed2-411e-42b7-8dd7-71b5ac2c8aab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15799"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(feat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "202d48a8-ecfa-46d4-8dbf-f87006c78ea8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(81.65508021390374, 183.48992184286303)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(list(ua.values()))/len(ua), sum(list(ua_dist.values()))/len(ua_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bf236ff4-d107-44aa-b402-6fd888999a8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 599)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min(list(ua_dist.values())), max(list(ua_dist.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9fe0562d-e050-45f0-a97c-4b35089824cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2079, 2783)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len([item for item in list(ua_dist.keys()) if item.type.endswith('same')]), len([item for item in list(ua_dist.keys()) if item.type.endswith('opp')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6743c4d6-aaf1-477f-85b1-c7f1630d542e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: ylabel='Frequency'>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkQAAAGdCAYAAADzOWwgAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjYsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvq6yFwwAAAAlwSFlzAAAPYQAAD2EBqD+naQAALUlJREFUeJzt3Xt0lNW9//HPhJAQLpNwMbcaIEdRiFxUojEFPfWQQ5DUitAe0VgiZkHVREFAhVrwbhCPVLE01B4FXGJRzhGqKGgMCIoxQAC5iAEVCQiT0IZkSCwhl/37wx/TDqBCmMxM2O/XWs9azN575vnuveLMx2ee5xmHMcYIAADAYiGBLgAAACDQCEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOuFBrqA1qCpqUkHDhxQp06d5HA4Al0OAAA4DcYYHTlyRPHx8QoJ+eFjQASi03DgwAElJCQEugwAANAM+/bt0/nnn/+DYwhEp6FTp06SvltQp9MZ4GoAAMDpcLvdSkhI8HyO/xAC0Wk4/jWZ0+kkEAEA0MqczukunFQNAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYL3QQBeA1qnn1LcDXcIZ+3pmRqBLAAAEqYAeIVq7dq2uv/56xcfHy+FwaNmyZZ6++vp6PfDAA+rXr586dOig+Ph4jRkzRgcOHPB6jcrKSmVmZsrpdCoqKkrZ2dmqqanxGrN161ZdffXVateunRISEjRr1ix/TA8AALQSAQ1EtbW1GjBggObOnXtS37fffqtNmzZp+vTp2rRpk9544w2VlpbqF7/4hde4zMxM7dixQwUFBVq+fLnWrl2r8ePHe/rdbreGDh2qHj16qKSkRE8//bQefvhhvfDCCy0+PwAA0Do4jDEm0EVIksPh0NKlSzVixIjvHbNhwwZdeeWV2rt3r7p3766dO3cqKSlJGzZsUHJysiRp5cqVGj58uPbv36/4+Hjl5+frwQcflMvlUlhYmCRp6tSpWrZsmT7//PPTqs3tdisyMlLV1dVyOp1nPddzAV+ZAQCC3Zl8freqk6qrq6vlcDgUFRUlSSoqKlJUVJQnDElSWlqaQkJCVFxc7BlzzTXXeMKQJKWnp6u0tFSHDx8+5X7q6urkdru9NgAAcO5qNYHo6NGjeuCBB3TzzTd7Up7L5VJ0dLTXuNDQUHXp0kUul8szJiYmxmvM8cfHx5woLy9PkZGRni0hIcHX0wEAAEGkVQSi+vp6/dd//ZeMMcrPz2/x/U2bNk3V1dWebd++fS2+TwAAEDhBf9n98TC0d+9erVq1yus7wNjYWFVUVHiNb2hoUGVlpWJjYz1jysvLvcYcf3x8zInCw8MVHh7uy2kAAIAgFtRHiI6Hod27d+v9999X165dvfpTU1NVVVWlkpIST9uqVavU1NSklJQUz5i1a9eqvr7eM6agoEAXX3yxOnfu7J+JAACAoBbQQFRTU6MtW7Zoy5YtkqQ9e/Zoy5YtKisrU319vX75y19q48aNWrRokRobG+VyueRyuXTs2DFJUp8+fTRs2DCNGzdO69ev17p165Sbm6vRo0crPj5eknTLLbcoLCxM2dnZ2rFjh1577TU999xzmjRpUqCmDQAAgkxAL7v/4IMPdO21157UnpWVpYcffliJiYmnfN7q1av1s5/9TNJ3N2bMzc3VW2+9pZCQEI0aNUpz5sxRx44dPeO3bt2qnJwcbdiwQd26ddPdd9+tBx544LTr5LL7k3HZPQAg2J3J53fQ3IcomBGITkYgAgAEu3P2PkQAAAAtgUAEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9UIDXQCknlPfDnQJAABYjSNEAADAegENRGvXrtX111+v+Ph4ORwOLVu2zKvfGKMZM2YoLi5OERERSktL0+7du73GVFZWKjMzU06nU1FRUcrOzlZNTY3XmK1bt+rqq69Wu3btlJCQoFmzZrX01AAAQCsS0EBUW1urAQMGaO7cuafsnzVrlubMmaN58+apuLhYHTp0UHp6uo4ePeoZk5mZqR07dqigoEDLly/X2rVrNX78eE+/2+3W0KFD1aNHD5WUlOjpp5/Www8/rBdeeKHF5wcAAFoHhzHGBLoISXI4HFq6dKlGjBgh6bujQ/Hx8Zo8ebKmTJkiSaqurlZMTIwWLFig0aNHa+fOnUpKStKGDRuUnJwsSVq5cqWGDx+u/fv3Kz4+Xvn5+XrwwQflcrkUFhYmSZo6daqWLVumzz///LRqc7vdioyMVHV1tZxOp8/nzjlE/vH1zIxAlwAA8KMz+fwO2nOI9uzZI5fLpbS0NE9bZGSkUlJSVFRUJEkqKipSVFSUJwxJUlpamkJCQlRcXOwZc80113jCkCSlp6ertLRUhw8fPuW+6+rq5Ha7vTYAAHDuCtpA5HK5JEkxMTFe7TExMZ4+l8ul6Ohor/7Q0FB16dLFa8ypXuNf93GivLw8RUZGeraEhISznxAAAAhaQRuIAmnatGmqrq72bPv27Qt0SQAAoAUFbSCKjY2VJJWXl3u1l5eXe/piY2NVUVHh1d/Q0KDKykqvMad6jX/dx4nCw8PldDq9NgAAcO4K2kCUmJio2NhYFRYWetrcbreKi4uVmpoqSUpNTVVVVZVKSko8Y1atWqWmpialpKR4xqxdu1b19fWeMQUFBbr44ovVuXNnP80GAAAEs4AGopqaGm3ZskVbtmyR9N2J1Fu2bFFZWZkcDocmTpyoxx9/XG+++aa2bdumMWPGKD4+3nMlWp8+fTRs2DCNGzdO69ev17p165Sbm6vRo0crPj5eknTLLbcoLCxM2dnZ2rFjh1577TU999xzmjRpUoBmDQAAgk1Af7pj48aNuvbaaz2Pj4eUrKwsLViwQPfff79qa2s1fvx4VVVVafDgwVq5cqXatWvnec6iRYuUm5urIUOGKCQkRKNGjdKcOXM8/ZGRkXrvvfeUk5OjgQMHqlu3bpoxY4bXvYoAAIDdguY+RMGM+xCdG7gPEQDY5Zy4DxEAAIC/EIgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPVCA10A4C89p74d6BLO2NczMwJdAgBYgSNEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFgvqANRY2Ojpk+frsTEREVEROiCCy7QY489JmOMZ4wxRjNmzFBcXJwiIiKUlpam3bt3e71OZWWlMjMz5XQ6FRUVpezsbNXU1Ph7OgAAIEgFdSB66qmnlJ+frz/84Q/auXOnnnrqKc2aNUvPP/+8Z8ysWbM0Z84czZs3T8XFxerQoYPS09N19OhRz5jMzEzt2LFDBQUFWr58udauXavx48cHYkoAACAIOcy/Hm4JMj//+c8VExOjF1980dM2atQoRURE6JVXXpExRvHx8Zo8ebKmTJkiSaqurlZMTIwWLFig0aNHa+fOnUpKStKGDRuUnJwsSVq5cqWGDx+u/fv3Kz4+/kfrcLvdioyMVHV1tZxOp8/n2XPq2z5/TZwbvp6ZEegSAKDVOpPP76A+QvTTn/5UhYWF2rVrlyTp008/1UcffaTrrrtOkrRnzx65XC6lpaV5nhMZGamUlBQVFRVJkoqKihQVFeUJQ5KUlpamkJAQFRcXn3K/dXV1crvdXhsAADh3hQa6gB8ydepUud1u9e7dW23atFFjY6OeeOIJZWZmSpJcLpckKSYmxut5MTExnj6Xy6Xo6Giv/tDQUHXp0sUz5kR5eXl65JFHfD0dAAAQpIL6CNHrr7+uRYsW6dVXX9WmTZu0cOFC/fd//7cWLlzYovudNm2aqqurPdu+fftadH8AACCwgvoI0X333aepU6dq9OjRkqR+/fpp7969ysvLU1ZWlmJjYyVJ5eXliouL8zyvvLxcl156qSQpNjZWFRUVXq/b0NCgyspKz/NPFB4ervDw8BaYEQAACEZBfYTo22+/VUiId4lt2rRRU1OTJCkxMVGxsbEqLCz09LvdbhUXFys1NVWSlJqaqqqqKpWUlHjGrFq1Sk1NTUpJSfHDLAAAQLAL6iNE119/vZ544gl1795dl1xyiTZv3qzZs2fr9ttvlyQ5HA5NnDhRjz/+uHr16qXExERNnz5d8fHxGjFihCSpT58+GjZsmMaNG6d58+apvr5eubm5Gj169GldYQYAAM59QR2Inn/+eU2fPl133XWXKioqFB8fr9/85jeaMWOGZ8z999+v2tpajR8/XlVVVRo8eLBWrlypdu3aecYsWrRIubm5GjJkiEJCQjRq1CjNmTMnEFMCAABBKKjvQxQsuA8RAoX7EAFA850z9yECAADwBwIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1mhWIvvrqK1/XAQAAEDDNCkQXXnihrr32Wr3yyis6evSor2sCAADwq2YFok2bNql///6aNGmSYmNj9Zvf/Ebr16/3dW0AAAB+0axAdOmll+q5557TgQMH9NJLL+ngwYMaPHiw+vbtq9mzZ+vQoUO+rhMAAKDFnNVJ1aGhoRo5cqSWLFmip556Sl988YWmTJmihIQEjRkzRgcPHvRVnQAAAC3mrALRxo0bdddddykuLk6zZ8/WlClT9OWXX6qgoEAHDhzQDTfc4Ks6AQAAWkxoc540e/ZszZ8/X6WlpRo+fLhefvllDR8+XCEh3+WrxMRELViwQD179vRlrQAAAC2iWYEoPz9ft99+u2677TbFxcWdckx0dLRefPHFsyoOAADAH5oViHbv3v2jY8LCwpSVldWclwcAAPCrZp1DNH/+fC1ZsuSk9iVLlmjhwoVnXRQAAIA/NSsQ5eXlqVu3bie1R0dH68knnzzrogAAAPypWYGorKxMiYmJJ7X36NFDZWVlZ10UAACAPzUrEEVHR2vr1q0ntX/66afq2rXrWRcFAADgT80KRDfffLPuuecerV69Wo2NjWpsbNSqVas0YcIEjR492tc1AgAAtKhmXWX22GOP6euvv9aQIUMUGvrdSzQ1NWnMmDGcQwQAAFqdZgWisLAwvfbaa3rsscf06aefKiIiQv369VOPHj18XR8AAECLa1YgOu6iiy7SRRdd5KtaAAAAAqJZgaixsVELFixQYWGhKioq1NTU5NW/atUqnxQHAADgD80KRBMmTNCCBQuUkZGhvn37yuFw+LouAAAAv2lWIFq8eLFef/11DR8+3Nf1AAAA+F2zLrsPCwvThRde6OtaAAAAAqJZgWjy5Ml67rnnZIzxdT0AAAB+16yvzD766COtXr1aK1as0CWXXKK2bdt69b/xxhs+KQ4AAMAfmhWIoqKidOONN/q6FgAAgIBoViCaP3++r+sAAAAImGadQyRJDQ0Nev/99/WnP/1JR44ckSQdOHBANTU1PisOAADAH5p1hGjv3r0aNmyYysrKVFdXp//8z/9Up06d9NRTT6murk7z5s3zdZ0AAAAtpllHiCZMmKDk5GQdPnxYERERnvYbb7xRhYWFPisOAADAH5p1hOjDDz/Uxx9/rLCwMK/2nj176ptvvvFJYQAAAP7SrCNETU1NamxsPKl9//796tSp01kXBQAA4E/NCkRDhw7Vs88+63nscDhUU1Ojhx56iJ/zAAAArU6zvjJ75plnlJ6erqSkJB09elS33HKLdu/erW7duukvf/mLr2sEAABoUc0KROeff74+/fRTLV68WFu3blVNTY2ys7OVmZnpdZI1AABAa9CsQCRJoaGhuvXWW31ZCwAAQEA0KxC9/PLLP9g/ZsyYZhUDAAAQCM0KRBMmTPB6XF9fr2+//VZhYWFq3749gQgAALQqzbrK7PDhw15bTU2NSktLNXjwYE6qBgAArU6zf8vsRL169dLMmTNPOnoEAAAQ7HwWiKTvTrQ+cOCAL18SAACgxTXrHKI333zT67ExRgcPHtQf/vAHDRo0yCeFAQAA+EuzjhCNGDHCaxs5cqQefvhh9e/fXy+99JJPC/zmm2906623qmvXroqIiFC/fv20ceNGT78xRjNmzFBcXJwiIiKUlpam3bt3e71GZWWlMjMz5XQ6FRUVpezsbNXU1Pi0TgAA0Ho16whRU1OTr+s4pcOHD2vQoEG69tprtWLFCp133nnavXu3Onfu7Bkza9YszZkzRwsXLlRiYqKmT5+u9PR0ffbZZ2rXrp0kKTMzUwcPHlRBQYHq6+s1duxYjR8/Xq+++qpf5gEAAIKbwxhjAl3E95k6darWrVunDz/88JT9xhjFx8dr8uTJmjJliiSpurpaMTExWrBggUaPHq2dO3cqKSlJGzZsUHJysiRp5cqVGj58uPbv36/4+PgfrcPtdisyMlLV1dVyOp2+m+D/13Pq2z5/TZwbvp6ZEegSAKDVOpPP72YdIZo0adJpj509e3ZzdiHpu3OV0tPT9atf/Upr1qzRT37yE911110aN26cJGnPnj1yuVxKS0vzPCcyMlIpKSkqKirS6NGjVVRUpKioKE8YkqS0tDSFhISouLhYN95440n7raurU11dneex2+1u9hwAAEDwa1Yg2rx5szZv3qz6+npdfPHFkqRdu3apTZs2uvzyyz3jHA7HWRX31VdfKT8/X5MmTdJvf/tbbdiwQffcc4/CwsKUlZUll8slSYqJifF6XkxMjKfP5XIpOjraqz80NFRdunTxjDlRXl6eHnnkkbOqHQAAtB7NCkTXX3+9OnXqpIULF3rO5zl8+LDGjh2rq6++WpMnT/ZJcU1NTUpOTtaTTz4pSbrsssu0fft2zZs3T1lZWT7Zx6lMmzbN6yiY2+1WQkJCi+0PAAAEVrOuMnvmmWeUl5fndXJz586d9fjjj+uZZ57xWXFxcXFKSkryauvTp4/KysokSbGxsZKk8vJyrzHl5eWevtjYWFVUVHj1NzQ0qLKy0jPmROHh4XI6nV4bAAA4dzUrELndbh06dOik9kOHDunIkSNnXdRxgwYNUmlpqVfbrl271KNHD0lSYmKiYmNjVVhY6FVbcXGxUlNTJUmpqamqqqpSSUmJZ8yqVavU1NSklJQUn9UKAABar2YFohtvvFFjx47VG2+8of3792v//v36v//7P2VnZ2vkyJE+K+7ee+/VJ598oieffFJffPGFXn31Vb3wwgvKycmR9N05ShMnTtTjjz+uN998U9u2bdOYMWMUHx+vESNGSPruiNKwYcM0btw4rV+/XuvWrVNubq5Gjx59WleYAQCAc1+zziGaN2+epkyZoltuuUX19fXfvVBoqLKzs/X000/7rLgrrrhCS5cu1bRp0/Too48qMTFRzz77rDIzMz1j7r//ftXW1mr8+PGqqqrS4MGDtXLlSs89iCRp0aJFys3N1ZAhQxQSEqJRo0Zpzpw5PqsTAAC0bmd1H6La2lp9+eWXkqQLLrhAHTp08FlhwYT7ECFQuA8RADTfmXx+n9WPux48eFAHDx5Ur1691KFDBwXxPR4BAAC+V7MC0d///ncNGTJEF110kYYPH66DBw9KkrKzs312yT0AAIC/NCsQ3XvvvWrbtq3KysrUvn17T/tNN92klStX+qw4AAAAf2jWSdXvvfee3n33XZ1//vle7b169dLevXt9UhgAAIC/NOsIUW1trdeRoeMqKysVHh5+1kUBAAD4U7MC0dVXX62XX37Z89jhcKipqUmzZs3Stdde67PiAAAA/KFZX5nNmjVLQ4YM0caNG3Xs2DHdf//92rFjhyorK7Vu3Tpf1wgAANCimnWEqG/fvtq1a5cGDx6sG264QbW1tRo5cqQ2b96sCy64wNc1AgAAtKgzPkJUX1+vYcOGad68eXrwwQdboiYAAAC/OuMjRG3bttXWrVtbohYAAICAaNZXZrfeeqtefPFFX9cCAAAQEM06qbqhoUEvvfSS3n//fQ0cOPCk3zCbPXu2T4oDAADwhzMKRF999ZV69uyp7du36/LLL5ck7dq1y2uMw+HwXXUAAAB+cEaBqFevXjp48KBWr14t6buf6pgzZ45iYmJapDgAAAB/OKNziE78NfsVK1aotrbWpwUBAAD4W7NOqj7uxIAEAADQGp1RIHI4HCedI8Q5QwAAoLU7o3OIjDG67bbbPD/gevToUd1xxx0nXWX2xhtv+K5CwGI9p74d6BLO2NczMwJdAgCcsTMKRFlZWV6Pb731Vp8WAwAAEAhnFIjmz5/fUnUAAAAEzFmdVA0AAHAuIBABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1WlUgmjlzphwOhyZOnOhpO3r0qHJyctS1a1d17NhRo0aNUnl5udfzysrKlJGRofbt2ys6Olr33XefGhoa/Fw9AAAIVq0mEG3YsEF/+tOf1L9/f6/2e++9V2+99ZaWLFmiNWvW6MCBAxo5cqSnv7GxURkZGTp27Jg+/vhjLVy4UAsWLNCMGTP8PQUAABCkWkUgqqmpUWZmpv785z+rc+fOnvbq6mq9+OKLmj17tv7jP/5DAwcO1Pz58/Xxxx/rk08+kSS99957+uyzz/TKK6/o0ksv1XXXXafHHntMc+fO1bFjxwI1JQAAEERaRSDKyclRRkaG0tLSvNpLSkpUX1/v1d67d291795dRUVFkqSioiL169dPMTExnjHp6elyu93asWPHKfdXV1cnt9vttQEAgHNXaKAL+DGLFy/Wpk2btGHDhpP6XC6XwsLCFBUV5dUeExMjl8vlGfOvYeh4//G+U8nLy9Mjjzzig+oBAEBrENRHiPbt26cJEyZo0aJFateund/2O23aNFVXV3u2ffv2+W3fAADA/4I6EJWUlKiiokKXX365QkNDFRoaqjVr1mjOnDkKDQ1VTEyMjh07pqqqKq/nlZeXKzY2VpIUGxt70lVnxx8fH3Oi8PBwOZ1Orw0AAJy7gjoQDRkyRNu2bdOWLVs8W3JysjIzMz3/btu2rQoLCz3PKS0tVVlZmVJTUyVJqamp2rZtmyoqKjxjCgoK5HQ6lZSU5Pc5AQCA4BPU5xB16tRJffv29Wrr0KGDunbt6mnPzs7WpEmT1KVLFzmdTt19991KTU3VVVddJUkaOnSokpKS9Otf/1qzZs2Sy+XS7373O+Xk5Cg8PNzvcwIAAMEnqAPR6fj973+vkJAQjRo1SnV1dUpPT9cf//hHT3+bNm20fPly3XnnnUpNTVWHDh2UlZWlRx99NIBVAwCAYOIwxphAFxHs3G63IiMjVV1d3SLnE/Wc+rbPXxMIlK9nZgS6BACQdGaf30F9DhEAAIA/EIgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWCw10AQDOLT2nvh3oEs7Y1zMzAl0CgAAjEAEAcI7hf0zOHF+ZAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAev3YPwHr8MjgAjhABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFgvqANRXl6errjiCnXq1EnR0dEaMWKESktLvcYcPXpUOTk56tq1qzp27KhRo0apvLzca0xZWZkyMjLUvn17RUdH67777lNDQ4M/pwIAAIJYUAeiNWvWKCcnR5988okKCgpUX1+voUOHqra21jPm3nvv1VtvvaUlS5ZozZo1OnDggEaOHOnpb2xsVEZGho4dO6aPP/5YCxcu1IIFCzRjxoxATAkAAAQhhzHGBLqI03Xo0CFFR0drzZo1uuaaa1RdXa3zzjtPr776qn75y19Kkj7//HP16dNHRUVFuuqqq7RixQr9/Oc/14EDBxQTEyNJmjdvnh544AEdOnRIYWFhP7pft9utyMhIVVdXy+l0+nxerfGmcAACixsz4oe0xs+VlvibPpPP76A+QnSi6upqSVKXLl0kSSUlJaqvr1daWppnTO/evdW9e3cVFRVJkoqKitSvXz9PGJKk9PR0ud1u7dix45T7qaurk9vt9toAAMC5q9UEoqamJk2cOFGDBg1S3759JUkul0thYWGKioryGhsTEyOXy+UZ869h6Hj/8b5TycvLU2RkpGdLSEjw8WwAAEAwaTWBKCcnR9u3b9fixYtbfF/Tpk1TdXW1Z9u3b1+L7xMAAAROq/hx19zcXC1fvlxr167V+eef72mPjY3VsWPHVFVV5XWUqLy8XLGxsZ4x69ev93q941ehHR9zovDwcIWHh/t4FgAAIFgF9REiY4xyc3O1dOlSrVq1SomJiV79AwcOVNu2bVVYWOhpKy0tVVlZmVJTUyVJqamp2rZtmyoqKjxjCgoK5HQ6lZSU5J+JAACAoBbUR4hycnL06quv6q9//as6derkOecnMjJSERERioyMVHZ2tiZNmqQuXbrI6XTq7rvvVmpqqq666ipJ0tChQ5WUlKRf//rXmjVrllwul373u98pJyeHo0AAAEBSkAei/Px8SdLPfvYzr/b58+frtttukyT9/ve/V0hIiEaNGqW6ujqlp6frj3/8o2dsmzZttHz5ct15551KTU1Vhw4dlJWVpUcffdRf0wAAAEEuqAPR6dwiqV27dpo7d67mzp37vWN69Oihd955x5elAQCAc0hQn0MEAADgDwQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6QX2nagDAqfWc+nagS7DG1zMzAl0C/IAjRAAAwHoEIgAAYD0CEQAAsB7nEAEA8AM4X8sOHCECAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANYjEAEAAOsRiAAAgPUIRAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABYj0AEAACsRyACAADWIxABAADrEYgAAID1CEQAAMB6BCIAAGA9AhEAALAegQgAAFiPQAQAAKxHIAIAANazKhDNnTtXPXv2VLt27ZSSkqL169cHuiQAABAErAlEr732miZNmqSHHnpImzZt0oABA5Senq6KiopAlwYAAALMmkA0e/ZsjRs3TmPHjlVSUpLmzZun9u3b66WXXgp0aQAAIMBCA12APxw7dkwlJSWaNm2apy0kJERpaWkqKio6aXxdXZ3q6uo8j6urqyVJbre7Reprqvu2RV4XAIDWoiU+Y4+/pjHmR8daEYj+9re/qbGxUTExMV7tMTEx+vzzz08an5eXp0ceeeSk9oSEhBarEQAAm0U+23KvfeTIEUVGRv7gGCsC0ZmaNm2aJk2a5Hnc1NSkyspKde3aVQ6Hw2f7cbvdSkhI0L59++R0On32uucq1uv0sVanj7U6M6zX6WOtzkxLrJcxRkeOHFF8fPyPjrUiEHXr1k1t2rRReXm5V3t5ebliY2NPGh8eHq7w8HCvtqioqBarz+l08h/LGWC9Th9rdfpYqzPDep0+1urM+Hq9fuzI0HFWnFQdFhamgQMHqrCw0NPW1NSkwsJCpaamBrAyAAAQDKw4QiRJkyZNUlZWlpKTk3XllVfq2WefVW1trcaOHRvo0gAAQIBZE4huuukmHTp0SDNmzJDL5dKll16qlStXnnSitT+Fh4froYceOunrOZwa63X6WKvTx1qdGdbr9LFWZybQ6+Uwp3MtGgAAwDnMinOIAAAAfgiBCAAAWI9ABAAArEcgAgAA1iMQBdDcuXPVs2dPtWvXTikpKVq/fn2gS/K7tWvX6vrrr1d8fLwcDoeWLVvm1W+M0YwZMxQXF6eIiAilpaVp9+7dXmMqKyuVmZkpp9OpqKgoZWdnq6amxo+z8I+8vDxdccUV6tSpk6KjozVixAiVlpZ6jTl69KhycnLUtWtXdezYUaNGjTrphqRlZWXKyMhQ+/btFR0drfvuu08NDQ3+nEqLy8/PV//+/T03eEtNTdWKFSs8/azT95s5c6YcDocmTpzoaWO9/unhhx+Ww+Hw2nr37u3pZ628ffPNN7r11lvVtWtXRUREqF+/ftq4caOnP6je4w0CYvHixSYsLMy89NJLZseOHWbcuHEmKirKlJeXB7o0v3rnnXfMgw8+aN544w0jySxdutSrf+bMmSYyMtIsW7bMfPrpp+YXv/iFSUxMNP/4xz88Y4YNG2YGDBhgPvnkE/Phhx+aCy+80Nx8881+nknLS09PN/Pnzzfbt283W7ZsMcOHDzfdu3c3NTU1njF33HGHSUhIMIWFhWbjxo3mqquuMj/96U89/Q0NDaZv374mLS3NbN682bzzzjumW7duZtq0aYGYUot58803zdtvv2127dplSktLzW9/+1vTtm1bs337dmMM6/R91q9fb3r27Gn69+9vJkyY4Glnvf7poYceMpdccok5ePCgZzt06JCnn7X6p8rKStOjRw9z2223meLiYvPVV1+Zd99913zxxReeMcH0Hk8gCpArr7zS5OTkeB43Njaa+Ph4k5eXF8CqAuvEQNTU1GRiY2PN008/7Wmrqqoy4eHh5i9/+YsxxpjPPvvMSDIbNmzwjFmxYoVxOBzmm2++8VvtgVBRUWEkmTVr1hhjvlubtm3bmiVLlnjG7Ny500gyRUVFxpjvAmhISIhxuVyeMfn5+cbpdJq6ujr/TsDPOnfubP7nf/6HdfoeR44cMb169TIFBQXm3//93z2BiPXy9tBDD5kBAwacso+18vbAAw+YwYMHf29/sL3H85VZABw7dkwlJSVKS0vztIWEhCgtLU1FRUUBrCy47NmzRy6Xy2udIiMjlZKS4lmnoqIiRUVFKTk52TMmLS1NISEhKi4u9nvN/lRdXS1J6tKliySppKRE9fX1XuvVu3dvde/e3Wu9+vXr53VD0vT0dLndbu3YscOP1ftPY2OjFi9erNraWqWmprJO3yMnJ0cZGRle6yLxd3Uqu3fvVnx8vP7t3/5NmZmZKisrk8RanejNN99UcnKyfvWrXyk6OlqXXXaZ/vznP3v6g+09nkAUAH/729/U2Nh40l2yY2Ji5HK5AlRV8Dm+Fj+0Ti6XS9HR0V79oaGh6tKlyzm9lk1NTZo4caIGDRqkvn37SvpuLcLCwk76IeIT1+tU63m871yybds2dezYUeHh4brjjju0dOlSJSUlsU6nsHjxYm3atEl5eXkn9bFe3lJSUrRgwQKtXLlS+fn52rNnj66++modOXKEtTrBV199pfz8fPXq1Uvvvvuu7rzzTt1zzz1auHChpOB7j7fmpzuAc0lOTo62b9+ujz76KNClBK2LL75YW7ZsUXV1tf73f/9XWVlZWrNmTaDLCjr79u3ThAkTVFBQoHbt2gW6nKB33XXXef7dv39/paSkqEePHnr99dcVERERwMqCT1NTk5KTk/Xkk09Kki677DJt375d8+bNU1ZWVoCrOxlHiAKgW7duatOmzUlXHpSXlys2NjZAVQWf42vxQ+sUGxuriooKr/6GhgZVVlaes2uZm5ur5cuXa/Xq1Tr//PM97bGxsTp27Jiqqqq8xp+4Xqdaz+N955KwsDBdeOGFGjhwoPLy8jRgwAA999xzrNMJSkpKVFFRocsvv1yhoaEKDQ3VmjVrNGfOHIWGhiomJob1+gFRUVG66KKL9MUXX/C3dYK4uDglJSV5tfXp08fzFWOwvccTiAIgLCxMAwcOVGFhoaetqalJhYWFSk1NDWBlwSUxMVGxsbFe6+R2u1VcXOxZp9TUVFVVVamkpMQzZtWqVWpqalJKSorfa25Jxhjl5uZq6dKlWrVqlRITE736Bw4cqLZt23qtV2lpqcrKyrzWa9u2bV5vMAUFBXI6nSe9cZ1rmpqaVFdXxzqdYMiQIdq2bZu2bNni2ZKTk5WZmen5N+v1/WpqavTll18qLi6Ov60TDBo06KRbg+zatUs9evSQFITv8T49RRunbfHixSY8PNwsWLDAfPbZZ2b8+PEmKirK68oDGxw5csRs3rzZbN682Ugys2fPNps3bzZ79+41xnx3SWZUVJT561//arZu3WpuuOGGU16Sedlll5ni4mLz0UcfmV69ep2Tl93feeedJjIy0nzwwQdel/x+++23njF33HGH6d69u1m1apXZuHGjSU1NNampqZ7+45f8Dh061GzZssWsXLnSnHfeeefcJb9Tp041a9asMXv27DFbt241U6dONQ6Hw7z33nvGGNbpx/zrVWbGsF7/avLkyeaDDz4we/bsMevWrTNpaWmmW7dupqKiwhjDWv2r9evXm9DQUPPEE0+Y3bt3m0WLFpn27dubV155xTMmmN7jCUQB9Pzzz5vu3bubsLAwc+WVV5pPPvkk0CX53erVq42kk7asrCxjzHeXZU6fPt3ExMSY8PBwM2TIEFNaWur1Gn//+9/NzTffbDp27GicTqcZO3asOXLkSABm07JOtU6SzPz58z1j/vGPf5i77rrLdO7c2bRv397ceOON5uDBg16v8/XXX5vrrrvOREREmG7dupnJkyeb+vp6P8+mZd1+++2mR48eJiwszJx33nlmyJAhnjBkDOv0Y04MRKzXP910000mLi7OhIWFmZ/85Cfmpptu8rqvDmvl7a233jJ9+/Y14eHhpnfv3uaFF17w6g+m93iHMcb49pgTAABA68I5RAAAwHoEIgAAYD0CEQAAsB6BCAAAWI9ABAAArEcgAgAA1iMQAQAA6xGIAACA9QhEAADAegQiAABgPQIRAACwHoEIAABY7/8BvX29ZcvkdfkAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pd.Series(ua_dist).plot.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "54d2dceb-089e-46a4-a87b-23522b5476bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(418.7515786506267)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.Series(ua_dist).mean() + 1.5*pd.Series(ua_dist).std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "558168c9-8964-4e60-bbde-218219dfbac8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8b1c20e9-6fef-46fc-aa59-194fb612d639",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ua, ua_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9603b8a7-10fa-4e29-9f84-a2bf550c81f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ua[Feat(chr=1.0, start=335, end=649, strand='+', type='gene_right')], ua_dist[Feat(chr=1.0, start=335, end=649, strand='+', type='gene_right')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "89a62519-4c1a-43d0-a778-210457819cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in ua:\n",
    "    if key.strand == '+' and key.type.endswith('same'):\n",
    "        key2 = Feat(key.chr, key.end, key.end+ua_dist[key], key.strand, key.type)\n",
    "    if key.strand == '+' and key.type.endswith('opp'):\n",
    "        key2 = Feat(key.chr, key.start-ua_dist[key], key.start, '-', key.type)\n",
    "    if key.strand == '-' and key.type.endswith('same'):\n",
    "        key2 = Feat(key.chr, key.start-ua_dist[key], key.start, key.strand, key.type)\n",
    "    if key.strand == '-' and key.type.endswith('opp'):\n",
    "        key2 = Feat(key.chr, key.end, key.end+ua_dist[key], '+', key.type)\n",
    "        \n",
    "    d[key2] = ua[key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "04a2dac2-d20e-4c71-8a0e-a95404698f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5248bbea-b7f6-4465-a068-6736a1dcae64",
   "metadata": {},
   "outputs": [],
   "source": [
    "expand = defaultdict(int)\n",
    "for row in d:\n",
    "    for i in range(row.start, row.end+1):\n",
    "        expand[(row.chr, i, row.strand)] += d[row] /(row.end-row.start+1)\n",
    "        # expand_type[(row.chr, i, row.strand)].add(row.type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d3a88037-e3dc-4d32-b370-a29335f2c9da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9368</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9369</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9370</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9371</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9372</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr   pos strand     reads\n",
       "0  1.0  9368      +  0.317597\n",
       "1  1.0  9369      +  0.317597\n",
       "2  1.0  9370      +  0.317597\n",
       "3  1.0  9371      +  0.317597\n",
       "4  1.0  9372      +  0.317597"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame([expand])\n",
    "df = df.transpose()\n",
    "df = df.rename(columns={0: \"reads\"})\n",
    "df = df.reset_index(names = \"og_index\")\n",
    "df = pd.concat([pd.DataFrame(df[\"og_index\"].tolist(), columns = [\"chr\", \"pos\", \"strand\"]), df[[\"reads\"]]], axis=1) \n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "87b8f797-d685-4df2-a845-38dc41abe6b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"/outs/net_processed.txt\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "0767ec51-b9e4-4283-bcc8-44058d0d5dac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>230218</td>\n",
       "      <td>.</td>\n",
       "      <td>chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>801</td>\n",
       "      <td>-</td>\n",
       "      <td>telomere</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>337</td>\n",
       "      <td>801</td>\n",
       "      <td>-</td>\n",
       "      <td>X_element</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>63</td>\n",
       "      <td>336</td>\n",
       "      <td>-</td>\n",
       "      <td>X_element_combinatorial_repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>-</td>\n",
       "      <td>telomeric_repeat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr  start     end strand                            type\n",
       "0  1.0      1  230218      .                      chromosome\n",
       "1  1.0      1     801      -                        telomere\n",
       "2  1.0    337     801      -                       X_element\n",
       "3  1.0     63     336      -  X_element_combinatorial_repeat\n",
       "4  1.0      1      62      -                telomeric_repeat"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allfeat = pd.read_table(\"/outs/all_features.txt\", sep=' ')\n",
    "allfeat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b0a12425-1e23-4128-8c0c-d7b9015074c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "allfeat = allfeat[allfeat[\"type\"] != \"chromosome\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b5a4cc8a-1639-4f7e-9019-56ca845336ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "allpos = defaultdict(list)\n",
    "for row in allfeat.itertuples(index=False):\n",
    "    for i in range(row.start, row.end+1):\n",
    "        allpos[(row.chr, i, row.strand)].append(row.type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ce3116d2-e566-4840-a79d-9862a5b0365d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([(1.0, 1, '-'), (1.0, 2, '-'), (1.0, 3, '-'), (1.0, 4, '-'), (1.0, 5, '-')],\n",
       " [['telomere', 'telomeric_repeat', 'telomere'],\n",
       "  ['telomere', 'telomeric_repeat', 'telomere'],\n",
       "  ['telomere', 'telomeric_repeat', 'telomere'],\n",
       "  ['telomere', 'telomeric_repeat', 'telomere'],\n",
       "  ['telomere', 'telomeric_repeat', 'telomere']])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(allpos.keys())[:5], list(allpos.values())[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "22220fbe-6d05-4142-83b1-9716d359bf0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr  pos strand                                   types\n",
       "0  1.0    1      -  [telomere, telomeric_repeat, telomere]\n",
       "1  1.0    2      -  [telomere, telomeric_repeat, telomere]\n",
       "2  1.0    3      -  [telomere, telomeric_repeat, telomere]\n",
       "3  1.0    4      -  [telomere, telomeric_repeat, telomere]\n",
       "4  1.0    5      -  [telomere, telomeric_repeat, telomere]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.DataFrame([allpos])\n",
    "df2 = df2.transpose()\n",
    "df2 = df2.rename(columns={0: \"types\"})\n",
    "df2 = df2.reset_index(names = \"og_index\")\n",
    "df2 = pd.concat([pd.DataFrame(df2[\"og_index\"].tolist(), columns = [\"chr\", \"pos\", \"strand\"]), df2[[\"types\"]]], axis=1) \n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "fd3c0993-75ad-42c2-ad98-36408221b329",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14051387</th>\n",
       "      <td>16.0</td>\n",
       "      <td>931342</td>\n",
       "      <td>+</td>\n",
       "      <td>[other]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14051388</th>\n",
       "      <td>16.0</td>\n",
       "      <td>931343</td>\n",
       "      <td>+</td>\n",
       "      <td>[other]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14051389</th>\n",
       "      <td>16.0</td>\n",
       "      <td>931344</td>\n",
       "      <td>+</td>\n",
       "      <td>[other]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14051390</th>\n",
       "      <td>16.0</td>\n",
       "      <td>931345</td>\n",
       "      <td>+</td>\n",
       "      <td>[other]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14051391</th>\n",
       "      <td>16.0</td>\n",
       "      <td>931346</td>\n",
       "      <td>+</td>\n",
       "      <td>[other]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr     pos strand    types\n",
       "14051387  16.0  931342      +  [other]\n",
       "14051388  16.0  931343      +  [other]\n",
       "14051389  16.0  931344      +  [other]\n",
       "14051390  16.0  931345      +  [other]\n",
       "14051391  16.0  931346      +  [other]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f68ef10b-e1d9-4aeb-b8a1-2b23323a901f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2[\"flag\"] = df2[\"types\"].apply(lambda item: \"tRNA\" in item or \"rRNA\" in item or \"transcript\" in item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0e20f253-5379-4d7b-a8e6-79bfc59dc181",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>types</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>127524</th>\n",
       "      <td>1.0</td>\n",
       "      <td>139152</td>\n",
       "      <td>+</td>\n",
       "      <td>[tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127525</th>\n",
       "      <td>1.0</td>\n",
       "      <td>139153</td>\n",
       "      <td>+</td>\n",
       "      <td>[tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127526</th>\n",
       "      <td>1.0</td>\n",
       "      <td>139154</td>\n",
       "      <td>+</td>\n",
       "      <td>[tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127527</th>\n",
       "      <td>1.0</td>\n",
       "      <td>139155</td>\n",
       "      <td>+</td>\n",
       "      <td>[tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127528</th>\n",
       "      <td>1.0</td>\n",
       "      <td>139156</td>\n",
       "      <td>+</td>\n",
       "      <td>[tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     pos strand                                              types  \\\n",
       "127524  1.0  139152      +  [tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...   \n",
       "127525  1.0  139153      +  [tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...   \n",
       "127526  1.0  139154      +  [tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...   \n",
       "127527  1.0  139155      +  [tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...   \n",
       "127528  1.0  139156      +  [tRNA_gene, noncoding_exon, tRNA, gene, mRNA, ...   \n",
       "\n",
       "        flag  \n",
       "127524  True  \n",
       "127525  True  \n",
       "127526  True  \n",
       "127527  True  \n",
       "127528  True  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[df2[\"flag\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7addf566-1791-4b60-95bc-e80dcc92181d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d7521bb3-c6e4-46b0-a35b-f4b8b7c941b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>types</th>\n",
       "      <th>flag</th>\n",
       "      <th>coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr  pos strand                                   types   flag  coding\n",
       "0  1.0    1      -  [telomere, telomeric_repeat, telomere]  False   False\n",
       "1  1.0    2      -  [telomere, telomeric_repeat, telomere]  False   False\n",
       "2  1.0    3      -  [telomere, telomeric_repeat, telomere]  False   False\n",
       "3  1.0    4      -  [telomere, telomeric_repeat, telomere]  False   False\n",
       "4  1.0    5      -  [telomere, telomeric_repeat, telomere]  False   False"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[\"coding\"] = df2[\"types\"].apply(lambda item: \"gene\" in item or \"CDS\" in item or \"mRNA\" in item)\n",
    "# df2[\"coding\"] = df2[\"types\"].apply(lambda item: \"mRNA\" in item or \"CDS\" in item)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0c17d6a7-1f0c-489b-89b9-92a9f0738047",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>types</th>\n",
       "      <th>flag</th>\n",
       "      <th>coding</th>\n",
       "      <th>features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr  pos strand                                   types   flag  coding  \\\n",
       "0  1.0    1      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "1  1.0    2      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "2  1.0    3      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "3  1.0    4      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "4  1.0    5      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "\n",
       "   features  \n",
       "0      True  \n",
       "1      True  \n",
       "2      True  \n",
       "3      True  \n",
       "4      True  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[\"features\"] = True\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8d86fb5b-797d-4abb-8798-2208acc163eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2[\"annotated_types\"] = df2[\"types\"].apply(lambda item: [i for i in item if i not in [\"SUTs\", \"CUTs\", \"XUTs\", \"other\"]])\n",
    "# df2[\"annotated_types\"] = df2[\"types\"].apply(lambda item: [i for i in item if i not in [\"SUTs\", \"CUTs\", \"XUTs\", \"other\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7432b6b1-87c9-4736-b643-6b84ae14e326",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>types</th>\n",
       "      <th>flag</th>\n",
       "      <th>coding</th>\n",
       "      <th>features</th>\n",
       "      <th>annotated_types</th>\n",
       "      <th>annotated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr  pos strand                                   types   flag  coding  \\\n",
       "0  1.0    1      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "1  1.0    2      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "2  1.0    3      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "3  1.0    4      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "4  1.0    5      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "\n",
       "   features                         annotated_types  annotated  \n",
       "0      True  [telomere, telomeric_repeat, telomere]       True  \n",
       "1      True  [telomere, telomeric_repeat, telomere]       True  \n",
       "2      True  [telomere, telomeric_repeat, telomere]       True  \n",
       "3      True  [telomere, telomeric_repeat, telomere]       True  \n",
       "4      True  [telomere, telomeric_repeat, telomere]       True  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[\"annotated\"] = df2[\"annotated_types\"].apply(lambda item: len(item) > 0)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "cb823509-250e-4533-891c-fe4925d8569b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>types</th>\n",
       "      <th>flag</th>\n",
       "      <th>coding</th>\n",
       "      <th>features</th>\n",
       "      <th>annotated_types</th>\n",
       "      <th>annotated</th>\n",
       "      <th>usefulish</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr  pos strand                                   types   flag  coding  \\\n",
       "0  1.0    1      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "1  1.0    2      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "2  1.0    3      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "3  1.0    4      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "4  1.0    5      -  [telomere, telomeric_repeat, telomere]  False   False   \n",
       "\n",
       "   features                         annotated_types  annotated  usefulish  \n",
       "0      True  [telomere, telomeric_repeat, telomere]       True      False  \n",
       "1      True  [telomere, telomeric_repeat, telomere]       True      False  \n",
       "2      True  [telomere, telomeric_repeat, telomere]       True      False  \n",
       "3      True  [telomere, telomeric_repeat, telomere]       True      False  \n",
       "4      True  [telomere, telomeric_repeat, telomere]       True      False  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2[\"usefulish\"] = df2[\"types\"].apply(lambda item: len(set(item).intersection(set([\"gene\", \"CDS\", \"mRNA\", \"ncRNA_gene\", \"tRNA_gene\", \"ncRNA\", \"tRNA\",\\\n",
    "                                                                           \"snoRNA_gene\", \"snoRNA\", \"telomerase_RNA_gene\", \"telomerase_RNA\", \\\n",
    "                                                                           \"snRNA_gene\", \"snRNA\"]))) > 0)\n",
    "# df2[\"usefulish\"] = df2[\"types\"].apply(lambda item: len(set(item).intersection(set([\"origin_of_replication\", \"gene\", \"mRNA\", \"exon\", \\\n",
    "#                                                                                    \"CDS\", \"regulatory_region\", \"ncRNA\", \"tRNA\", \"snoRNA\", \\\n",
    "#                                                                                    \"telomerase_RNA\", \"sequence_feature\", \"snRNA\", \"RNase_P_RNA\", \\\n",
    "#                                                                                    \"SRP_RNA\", \"transcript\", \"rRNA\", \"RNase_MRP_RNA\"]))) > 0)\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e74d6e10-f134-4869-b639-9bfb0f2a3057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>types</th>\n",
       "      <th>flag</th>\n",
       "      <th>coding</th>\n",
       "      <th>features</th>\n",
       "      <th>annotated_types</th>\n",
       "      <th>annotated</th>\n",
       "      <th>usefulish</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">1.0</th>\n",
       "      <th>1</th>\n",
       "      <th>-</th>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <th>-</th>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <th>-</th>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <th>-</th>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <th>-</th>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>[telomere, telomeric_repeat, telomere]</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 types   flag  coding  \\\n",
       "chr pos strand                                                          \n",
       "1.0 1   -       [telomere, telomeric_repeat, telomere]  False   False   \n",
       "    2   -       [telomere, telomeric_repeat, telomere]  False   False   \n",
       "    3   -       [telomere, telomeric_repeat, telomere]  False   False   \n",
       "    4   -       [telomere, telomeric_repeat, telomere]  False   False   \n",
       "    5   -       [telomere, telomeric_repeat, telomere]  False   False   \n",
       "\n",
       "                features                         annotated_types  annotated  \\\n",
       "chr pos strand                                                                \n",
       "1.0 1   -           True  [telomere, telomeric_repeat, telomere]       True   \n",
       "    2   -           True  [telomere, telomeric_repeat, telomere]       True   \n",
       "    3   -           True  [telomere, telomeric_repeat, telomere]       True   \n",
       "    4   -           True  [telomere, telomeric_repeat, telomere]       True   \n",
       "    5   -           True  [telomere, telomeric_repeat, telomere]       True   \n",
       "\n",
       "                usefulish  \n",
       "chr pos strand             \n",
       "1.0 1   -           False  \n",
       "    2   -           False  \n",
       "    3   -           False  \n",
       "    4   -           False  \n",
       "    5   -           False  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df2.set_index([\"chr\", \"pos\", \"strand\"])\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "cf9b2376-e633-4bc9-a5b7-edbba8bc5161",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.join(df2[[\"coding\", \"annotated\", \"features\", \"usefulish\", \"types\", \"flag\"]], on=[\"chr\", \"pos\", \"strand\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "780e754f-d88d-49c9-8812-cd4d8a6b9139",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>reads</th>\n",
       "      <th>coding</th>\n",
       "      <th>annotated</th>\n",
       "      <th>features</th>\n",
       "      <th>usefulish</th>\n",
       "      <th>types</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9368</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9369</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9370</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9371</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9372</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr   pos strand     reads coding annotated features usefulish   types  \\\n",
       "0  1.0  9368      +  0.317597  False     False     True     False  [SUTs]   \n",
       "1  1.0  9369      +  0.317597  False     False     True     False  [SUTs]   \n",
       "2  1.0  9370      +  0.317597  False     False     True     False  [SUTs]   \n",
       "3  1.0  9371      +  0.317597  False     False     True     False  [SUTs]   \n",
       "4  1.0  9372      +  0.317597  False     False     True     False  [SUTs]   \n",
       "\n",
       "    flag  \n",
       "0  False  \n",
       "1  False  \n",
       "2  False  \n",
       "3  False  \n",
       "4  False  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "63d8e0c4-4b3f-4a5a-ad38-2f2c09da23e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.to_csv(\"clean/big_processed.txt\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3c9a0464-3290-4afa-b8b3-a9b83da966a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>reads</th>\n",
       "      <th>coding</th>\n",
       "      <th>annotated</th>\n",
       "      <th>features</th>\n",
       "      <th>usefulish</th>\n",
       "      <th>types</th>\n",
       "      <th>flag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9368</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9369</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9370</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9371</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>9372</td>\n",
       "      <td>+</td>\n",
       "      <td>0.317597</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>[SUTs]</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr   pos strand     reads coding annotated features usefulish   types  \\\n",
       "0  1.0  9368      +  0.317597  False     False     True     False  [SUTs]   \n",
       "1  1.0  9369      +  0.317597  False     False     True     False  [SUTs]   \n",
       "2  1.0  9370      +  0.317597  False     False     True     False  [SUTs]   \n",
       "3  1.0  9371      +  0.317597  False     False     True     False  [SUTs]   \n",
       "4  1.0  9372      +  0.317597  False     False     True     False  [SUTs]   \n",
       "\n",
       "    flag  \n",
       "0  False  \n",
       "1  False  \n",
       "2  False  \n",
       "3  False  \n",
       "4  False  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df = pd.read_csv(\"clean/big_processed.txt\", low_memory=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f4331874-008f-4269-9ad7-ba6d4c01070e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_168345/2414254382.py:1: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df[\"coding\"] = df[\"coding\"].fillna(False)\n",
      "/tmp/ipykernel_168345/2414254382.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df[\"annotated\"] = df[\"annotated\"].fillna(False)\n",
      "/tmp/ipykernel_168345/2414254382.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df[\"features\"] = df[\"features\"].fillna(False)\n",
      "/tmp/ipykernel_168345/2414254382.py:4: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  df[\"usefulish\"] = df[\"usefulish\"].fillna(False)\n"
     ]
    }
   ],
   "source": [
    "df[\"coding\"] = df[\"coding\"].fillna(False)\n",
    "df[\"annotated\"] = df[\"annotated\"].fillna(False)\n",
    "df[\"features\"] = df[\"features\"].fillna(False)\n",
    "df[\"usefulish\"] = df[\"usefulish\"].fillna(False)\n",
    "# df[\"types\"] = \n",
    "df[\"types\"] = df[\"types\"].fillna(\"unannotated\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ed13205f-0e09-4712-9a8f-02d9f51121be",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"/outs/big_processed.txt\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "85d52942-719d-4672-8a30-92c00443535b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1706.5827240943909"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t1 = time.time()\n",
    "t1-t0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed48cd04-3345-449f-ae97-46473e6aae8d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f004d17a-438e-495b-b69a-de25d357b68f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a084c9ac-cd8a-45f9-93c2-a7ae49324b98",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
