---
title: "More Analysis"
output: html_document
---

```{r}
library(tidyverse)
library(MASS)

print("loading data")
d <- read.delim("/outs/the_greater_big_data_file_bigger.txt", header = TRUE, sep = " ")
print("loaded data. running models. breathe.")

d$rep.fitted <- (d$rep.fitted.dip + d$rep.fitted.hap) / 2
d$collision <- (d$collision.dip + d$collision.hap) / 2

# d <- d %>% filter(transcripts < 15)
d <- d %>% filter(transcripts < 2.5e4)

d2 <- d %>% filter(!is.na(length))

head(d)
```


```{r}
mod <- glm(SNM ~ ref * tr.sign + rep.fitted, d2, family = "binomial")
summary(mod)
```



Call:
glm(formula = SNM ~ base.type + base.type:annotated + rep.fitted + 
    rep.fitted:annotated, family = "binomial", data = d)

Coefficients:
                            Estimate Std. Error z value Pr(>|z|)    
(Intercept)                  -9.3176     0.3225 -28.894  < 2e-16 ***
base.typeCG                   0.7936     0.1425   5.568 2.58e-08 ***
base.typediNT                 0.4871     0.2773   1.757  0.07900 .  
rep.fitted                    0.5095     0.2179   2.338  0.01938 *  
base.typeAT:annotatedTRUE     0.7666     0.3372   2.274  0.02299 *  
base.typeCG:annotatedTRUE     0.8516     0.3345   2.546  0.01090 *  
base.typediNT:annotatedTRUE   1.6121     0.4134   3.900 9.63e-05 ***
annotatedTRUE:rep.fitted     -0.6880     0.2286  -3.009  0.00262 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 55085  on 12003428  degrees of freedom
Residual deviance: 54364  on 12003421  degrees of freedom
AIC: 54380

Number of Fisher Scoring iterations: 11


```{r}
d$annotated <- !is.na(d$length)

mod <- glm(SNM ~ (base.type + rep.fitted) * annotated, d, family = "binomial")
summary(mod)

mod <- glm(SNM ~ base.type + base.type:annotated + rep.fitted + rep.fitted:annotated, d, family = "binomial")
summary(mod)

mod <- glm(SNM ~ base.type + rep.fitted + rep.fitted:annotated, d, family = "binomial")
summary(mod)
```

Accounting for base type and replication timing, annotated regions have more mutations. Comment in discussion with potential reason why. 


```{r}
mod <- glm(
  formula = SNM ~ base.type + rep.fitted + protein + distance +
    collision + I(distance^2) + distance:collision + transcripts:length +
    protein:distance + base.type:distance + base.type:collision,
  family = "binomial", data = d2
)
summary(mod)
```

```{r}
library(car)
car::vif(mod)
```

seems ok??

```{r}
cooks_d <- cooks.distance(mod)
quantile(cooks_d)

quantile(cooks_d, 0.95)

# data.frame(cooks_d = cooks_d, chr=d2$chr, pos=d2$pos, transcripts=d2$transcripts, rep.fitted=d2$rep.fitted, length=d2$length, distance=d2$distance) %>%
#   filter(cooks_d>0.001) %>% view()

data.frame(cooks_d = cooks_d) %>%
  filter(cooks_d > 0.0001) %>%
  ggplot() +
  geom_histogram(aes(x = cooks_d))
```


```{r}
residual <- resid(mod)
# data.frame(residual=residual, chr=d2$chr, pos=d2$pos, transcripts=d2$transcripts, rep.fitted=d2$rep.fitted, length=d2$length, distance=d2$distance, cooks_d=cooks_d, SNM=d2$SNM) %>%
#   filter(residual>1) %>% view()


quantile(residual, 0.95)
quantile(d$transcripts)
```


These are all the mutations lol

```{r}
mean(d2$SNM)
var(d2$SNM)
```

```{r}
mod <- glm(
  formula = SNM ~ base.type + rep.fitted + protein + distance +
    collision + I(distance^2) + distance:collision + transcripts:length +
    protein:distance + base.type:distance + base.type:collision,
  family = "poisson", data = d2
)
summary(mod)
```


```{r}
probs <- dpois(0:1, lambda = mean(d2$SNM))
probs

comp <- 1 - sum(probs)
comp

chisq.test(x = c(sum(!d2$SNM), sum(d2$SNM), 0), p = c(probs, comp))
chisq.test(x = c(sum(!d2$SNM), sum(d2$SNM)), p = c(probs), rescale.p = TRUE)
```


```{r}
mod <- glm(
  formula = SNM ~ base.type + rep.fitted + protein + distance +
    collision + I(distance^2) + distance:collision + transcripts:length +
    protein:distance + base.type:distance + base.type:collision,
  family = "binomial", data = d2
)
summary(mod)
```

```{r}
d2 |>
  mutate(comp_res = coef(mod)["rep.fitted"] * rep.fitted + residuals(mod, type = "working")) |>
  ggplot(aes(x = rep.fitted, y = comp_res)) +
  geom_point() +
  geom_smooth(color = "red", method = "lm", linetype = 2, se = F) +
  geom_smooth(se = F)
```

```{r}
(d2 |>
  mutate(comp_res = coef(mod)["rep.fitted"] * rep.fitted + residuals(mod, type = "working")))$comp_res |> quantile(na.rm = TRUE)
```



```{r}
d2 |>
  mutate(comp_res = coef(mod)["rep.fitted"] * rep.fitted + residuals(mod, type = "working")) |>
  ggplot(aes(x = rep.fitted, y = comp_res)) +
  geom_point() +
  geom_smooth(color = "red", method = "lm", linetype = 2, se = F) +
  geom_smooth(se = F) +
  coord_cartesian(ylim = c(-1.560111, 1000))
```


```{r}
d2 |>
  mutate(comp_res = coef(mod)["rep.fitted"] * rep.fitted + residuals(mod, type = "working")) |>
  ggplot(aes(x = rep.fitted, y = comp_res)) +
  geom_point() +
  geom_smooth(color = "red", method = "lm", linetype = 2, se = F) +
  geom_smooth(se = F) +
  coord_cartesian(ylim = c(-1.560111, 20))
```

Yay this is gooooood!


```{r}
d2 |>
  mutate(comp_res = coef(mod)["distance"] * distance + residuals(mod, type = "working")) |>
  ggplot(aes(x = distance, y = comp_res)) +
  geom_point() +
  geom_smooth(color = "red", method = "lm", linetype = 2, se = F) +
  geom_smooth(se = F)
```

```{r}
(d2 |>
  mutate(comp_res = coef(mod)["distance"] * distance + residuals(mod, type = "working")))$comp_res |> quantile(na.rm = TRUE)
```


```{r}
d2 |>
  mutate(comp_res = coef(mod)["distance"] * distance + residuals(mod, type = "working")) |>
  ggplot(aes(x = distance, y = comp_res)) +
  geom_point() +
  geom_smooth(color = "red", method = "lm", linetype = 2, se = F) +
  geom_smooth(se = F) +
  coord_cartesian(ylim = c(-6, 20))
```
Diverging at higher distances. But how bad is bad?


```{r}
d2 <-
  d2 |>
  mutate(dffits = dffits(mod))

d2 %>%
  filter(abs(dffits) > 2 * sqrt(length(coef(mod)) / nobs(mod))) %>%
  view()

d2 |>
  mutate(
    obs_number = row_number(),
    large = ifelse(abs(dffits) > 2 * sqrt(length(coef(mod)) / nobs(mod)),
      "red", "black"
    )
  ) |>
  ggplot(aes(obs_number, dffits, color = large)) +
  geom_point() +
  geom_hline(yintercept = c(-1, 1) * 2 * sqrt(length(coef(mod)) / nobs(mod)), color = "red") +
  scale_color_identity()
```


```{r}
DescTools::PseudoR2(mod, which = "all")
```


Perhaps a comparison with a base model might help here?????

```{r}
mod0 <- glm(
  formula = SNM ~ base.type + rep.fitted,
  family = "binomial", data = d2
)
summary(mod0)
DescTools::PseudoR2(mod0, which = "all")
```

```{r}
mod <- glm(
  formula = SNM ~ base.type + rep.fitted + protein + distance +
    collision + I(distance^2) + base.type:collision + transcripts:length +
    protein:distance + base.type:distance + distance:collision,
  family = "binomial", data = d2
)
summary(mod)
```

```{r}
t <- d
t <- within(t, distance <- ifelse(is.na(distance), mean(d$distance, na.rm = TRUE), distance))
t <- within(t, length <- ifelse(is.na(length), mean(d$length, na.rm = TRUE), length))

t$predict_all <- predict(mod, newdata = t, type = "response")

q <- t %>%
  filter(is.na(d$distance)) %>%
  group_by(G = cut(predict_all, breaks = unique(quantile(predict_all, probs = seq(0, 1, by = 0.01))))) %>%
  summarize(predicted = mean(predict_all), observed = mean(SNM), count = n())
# q
nrow(q)
cor(q$predicted, q$observed)
q %>% ggplot() +
  geom_point(aes(y = observed, x = predicted)) +
  geom_smooth(aes(y = observed, x = predicted), method = "lm", se = FALSE, color = "red") +
  # geom_abline(slope=1, intercept=0, color="red") +
  xlab("Expected mutations in quantile") +
  ylab("Observed mutations in quantile") +
  theme(text = element_text(size = 18))
```


```{r}
t <- d
# t <- within(t, distance <- ifelse(is.na(distance), mean(d$distance, na.rm=TRUE), distance))
# t <- within(t, length <- ifelse(is.na(length), mean(d$length, na.rm=TRUE), length))

t <- within(t, distance <- ifelse(is.na(distance), 0, distance))
t <- within(t, length <- ifelse(is.na(length), 0, length))

t$predict_all <- predict(mod, newdata = t, type = "response")

q <- t %>%
  filter(is.na(d$distance)) %>%
  group_by(G = cut(predict_all, breaks = unique(quantile(predict_all, probs = seq(0, 1, by = 0.01))))) %>%
  summarize(predicted = mean(predict_all), observed = mean(SNM), count = n())
# q
nrow(q)
cor(q$predicted, q$observed)
q %>% ggplot() +
  geom_point(aes(y = observed, x = predicted)) +
  geom_smooth(aes(y = observed, x = predicted), method = "lm", se = FALSE, color = "red") +
  # geom_abline(slope=1, intercept=0, color="red") +
  xlab("Expected mutations in quantile") +
  ylab("Observed mutations in quantile") +
  theme(text = element_text(size = 18))
```




```{r}
q <- t %>%
  filter(!is.na(d$distance)) %>%
  group_by(G = cut(predict_all, breaks = unique(quantile(predict_all, probs = seq(0, 1, by = 0.01))))) %>%
  summarize(predicted = mean(predict_all), observed = mean(SNM), count = n())
# q
nrow(q)
cor(q$predicted, q$observed)
q %>% ggplot() +
  geom_point(aes(y = observed, x = predicted)) +
  geom_smooth(aes(y = observed, x = predicted), method = "lm", se = FALSE, color = "red") +
  # geom_abline(slope=1, intercept=0, color="red") +
  xlab("Expected mutations in quantile") +
  ylab("Observed mutations in quantile") +
  theme(text = element_text(size = 18))
```



```{r}
t <- d
t$distance <- mean(d$distance, na.rm = TRUE)
t$length <- mean(d$length, na.rm = TRUE)

t$predict_all <- predict(mod, newdata = t, type = "response")

q <- t %>%
  filter(!is.na(d$distance)) %>%
  group_by(G = cut(predict_all, breaks = unique(quantile(predict_all, probs = seq(0, 1, by = 0.01))))) %>%
  summarize(predicted = mean(predict_all), observed = mean(SNM), count = n())
# q
nrow(q)
cor(q$predicted, q$observed)
q %>% ggplot() +
  geom_point(aes(y = observed, x = predicted)) +
  geom_smooth(aes(y = observed, x = predicted), method = "lm", se = FALSE, color = "red") +
  # geom_abline(slope=1, intercept=0, color="red") +
  xlab("Expected mutations in quantile") +
  ylab("Observed mutations in quantile") +
  theme(text = element_text(size = 18))
```


```{r}
t <- d
t$distance <- 0
t$length <- 0
t$transcripts <- 0

t$predict_all <- predict(mod, newdata = t, type = "response")

q <- t %>%
  filter(!is.na(d$distance)) %>%
  group_by(G = cut(predict_all, breaks = unique(quantile(predict_all, probs = seq(0, 1, by = 0.01))))) %>%
  summarize(predicted = mean(predict_all), observed = mean(SNM), count = n())
# q
nrow(q)
cor(q$predicted, q$observed)
q %>% ggplot() +
  geom_point(aes(y = observed, x = predicted)) +
  geom_smooth(aes(y = observed, x = predicted), method = "lm", se = FALSE, color = "red") +
  # geom_abline(slope=1, intercept=0, color="red") +
  xlab("Expected mutations in quantile") +
  ylab("Observed mutations in quantile") +
  theme(text = element_text(size = 18))
```



```{r}
library(tidyverse)
print("loading data")
d <- read.delim("/outs/the_greater_big_data_file_bigger.txt", header = TRUE, sep = " ")
print("loaded data. running models. breathe.")

d$rep.fitted <- (d$rep.fitted.dip + d$rep.fitted.hap) / 2
d$collision <- (d$collision.dip + d$collision.hap) / 2

# d <- d %>% filter(transcripts < 15)
d <- d %>% filter(transcripts < 2.5e4)

d <- d %>% filter(!is.na(length))

head(d)
```

```{r}
mod <- glm(
  formula = SNM ~ base.type + rep.fitted + protein + distance +
    collision + I(distance^2) + base.type:collision + transcripts:length +
    protein:distance + base.type:distance + distance:collision,
  family = "binomial", data = d
)
summary(mod)
```


```{r}
mod2 <- glm(
  formula = SNM ~ base.type + rep.fitted + protein + distance +
    collision + I(distance^2) + base.type:collision + transcripts:length +
    protein:distance + base.type:distance + distance:collision + length,
  family = "binomial", data = d
)
summary(mod2)
```


```{r}
d$predicted <- predict(mod, newdata = d, type = "response")
```


```{r}
d %>% ggplot() +
  geom_density(aes(color=SNM, x=rep.fitted))
```



```{r}
m <- glm.nb(formula = SNM ~ base.type + rep.fitted + protein + distance +
  collision + I(distance^2) + base.type:collision + transcripts:length +
  protein:distance + base.type:distance + distance:collision, data = d)
summary(m)
```

```{r}
m2 <- glm(
  formula = SNM ~ base.type + rep.fitted + protein + distance +
    collision + I(distance^2) + base.type:collision + transcripts:length +
    protein:distance + base.type:distance + distance:collision,
  family = "poisson", data = d
)
summary(m2)
```

```{r}
pchisq(2 * (logLik(m) - logLik(m2)), df = 1, lower.tail = FALSE)
pchisq(2 * (logLik(m) - logLik(mod)), df = 1, lower.tail = FALSE)
```


Another sign that okay to poisson or binomial 

```{r}
pchisq(1, df = 16)

pchisq(926, df = 1)
```


```{r}
d$predicted <- predict(mod, newdata = d, type = "response")
```


Call:
glm(formula = SNM ~ base.type + rep.fitted + protein + distance + 
    collision + I(distance^2) + base.type:collision + transcripts:length + 
    protein:distance + base.type:distance + distance:collision, 
    family = "binomial", data = d)

Coefficients:
                          Estimate Std. Error z value Pr(>|z|)    
(Intercept)             -8.543e+00  1.442e-01 -59.245  < 2e-16 ***
base.typeCG              6.983e-01  7.600e-02   9.189  < 2e-16 ***
base.typediNT            1.449e+00  1.093e-01  13.259  < 2e-16 ***
rep.fitted              -2.793e-01  7.076e-02  -3.948 7.89e-05 ***
proteinTRUE              1.240e-01  1.040e-01   1.192 0.233378    
distance                -2.767e-04  8.039e-05  -3.442 0.000578 ***
collision               -2.241e-02  7.981e-02  -0.281 0.778818    
I(distance^2)           -1.718e-08  7.444e-09  -2.308 0.020985 *  
base.typeCG:collision    1.519e-01  8.745e-02   1.737 0.082323 .  
base.typediNT:collision -4.221e-02  1.273e-01  -0.332 0.740262    
transcripts:length       3.292e-09  1.866e-09   1.765 0.077603 .  
proteinTRUE:distance     3.076e-04  7.512e-05   4.095 4.23e-05 ***
base.typeCG:distance     8.446e-05  3.936e-05   2.146 0.031896 *  
base.typediNT:distance  -8.845e-05  6.486e-05  -1.364 0.172643    
distance:collision       6.455e-05  3.858e-05   1.673 0.094284 .  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 51148  on 11164203  degrees of freedom
Residual deviance: 50350  on 11164189  degrees of freedom
AIC: 50380

Number of Fisher Scoring iterations: 12


```{r}
q <- d %>%
  group_by(G = cut(predicted, breaks = unique(quantile(predicted, probs = seq(0, 1, by = 0.01))))) %>%
  summarize(predicted = mean(predicted), observed = mean(SNM), count = n())
# q
nrow(q)
cor(q$predicted, q$observed)
q %>% ggplot() +
  geom_point(aes(y = observed, x = predicted)) +
  geom_abline(slope=1, intercept=0, color="#C8102EFF", linetype="dashed") +
  geom_smooth(aes(y = observed, x = predicted), method = "lm", se = FALSE, color = "#C8102EFF") +
  # geom_abline(slope=1, intercept=0, color="red") +
  xlab("Expected mutations in quantile") +
  ylab("Observed mutations in quantile") +
  theme_bw(base_size=18)
  # theme(text = element_text(size = 18))
```

```{r}
n <- 10
dummy_df <- expand.grid(
  base.type = c("diNT", "CG", "AT"),
  rep.fitted = seq(min(d$rep.fitted), max(d$rep.fitted), length.out = n),
  protein = c(TRUE, FALSE),
  distance = seq(0, max(d$distance), length.out = n),
  collision = c(1, 0),
  transcripts = seq(0, max(d$transcripts), length.out = n),
  length = seq(0, max(d$length), length.out = n)
)
```


```{r}
dummy_df$predicted <- predict(mod, newdata = dummy_df, type = "response")
```

```{r}
library(DescTools)
```

```{r}
BinomCI(x=375, n=652353, method="wilson")
```

```{r}
library(paletteer)
```


```{r}
base.df <- data.frame(
  base.type = c("CpG", "C/G", "A/T"),
  sites = c(sum(d$base.type == "diNT"), sum(d$base.type == "CG"), sum(d$base.type == "AT")),
  mutations = c(
    sum(d[d$SNM, ]$base.type == "diNT"), sum(d[d$SNM, ]$base.type == "CG"),
    sum(d[d$SNM, ]$base.type == "AT")
  )
)

base.df$mutated_fraction <- base.df$mutations / base.df$sites
base.df[,c("est", "lower", "upper")] <- t(apply(base.df,1,function(r){BinomCI(as.numeric(r['mutations']), as.numeric(r['sites']), method="wilson")}))
base.df

base.df %>%
  ggplot() +
  geom_col(aes(x = base.type, y = mutated_fraction / 624598, fill = base.type), show.legend=FALSE) +
  geom_errorbar(aes(x = base.type, ymin = lower/624598, ymax = upper/624598), width=0.3) +
  # scale_fill_viridis_d(option = ) +
  # scale_fill_paletteer_d("LaCroixColoR::Pamplemousse") +
  scale_fill_paletteer_d("ggthemes::wsj_dem_rep") +
  xlab("Site Type") +
  ylab("Mutation Rate") +
  theme(text = element_text(size = 18))  +
  theme_bw(base_size=18)
```

```{r}
apply(base.df,1,function(r){BinomCI(as.numeric(r['mutations']), as.numeric(r['sites'][1]), method="wilson")})
```


```{r}
# 1. Enter data and calculate original mean/sd, n
cotanine <- d[d$SNM, ]$rep.fitted
x_bar <- mean(cotanine)
s <- sd(cotanine)
n <- length(cotanine)
se <- s / sqrt(n)

# Specify number of bootstrap samples, create a vector of that length
# to store t_hat values
B <- 10000
t_hat <- numeric(B)

# Bootstrap loop
for (i in 1:B) {
  # 2. Draw a SRS of size n from data
  x_star <- sample(cotanine, size = n, replace = T)

  # 3. Calculate resampled mean and sd
  x_bar_star <- mean(x_star)
  s_star <- sd(x_star)

  # 4. Calculate t_hat, and store it in vector
  t_hat[i] <- (x_bar_star - x_bar) / (s_star / sqrt(n))
}

# Optional: plot approximate sampling distribution
# hist(t_hat, main = "Approx. Sampling Distribution of t")

# Find upper and lower critical values of approx. distribution
t_lower <- quantile(t_hat, probs = 0.025, names = F)
t_upper <- quantile(t_hat, probs = 0.975, names = F)

# Calculate upper and lower bounds of CI
boot_ci_upper <- x_bar - t_lower * se
boot_ci_lower <- x_bar - t_upper * se

c(boot_ci_lower, boot_ci_upper)
```


```{r}
t.test(d[d$SNM, ]$rep.fitted)
```



```{r}
t.test(d[!d$SNM, ]$rep.fitted)
```


```{r}
library(cowplot)
```



```{r}
rep.df <- data.frame(
  site.type = c("SNM", "Non SNM"),
  rep.fitted = c(
    mean(d[d$SNM, ]$rep.fitted),
    mean(d[!d$SNM, ]$rep.fitted)
  ),
  lower = c(1.341558, 1.365688),
  upper = c(1.362010, 1.366016)
)

rep.plot <- rep.df %>%
  ggplot() +
  geom_point(aes(x = factor(site.type, levels = site.type), y = 2 - rep.fitted, color=site.type), show.legend = FALSE) +
  geom_errorbar(aes(x = factor(site.type, levels = site.type), y = 2 - rep.fitted, ymin = 2 - upper, ymax = 2 - lower, color=site.type), width=0.3, show.legend=FALSE) +
  scale_color_paletteer_d("nbapalettes::sixers_retro") +
  xlab("") +
  ylab("Replication timing") +
  theme_bw(base_size=12) +
  ylim(0.62, 0.67)

rep.plot
```

```{r}
p <- d %>% ggplot() + 
  # geom_density(aes(x=2-rep.fitted, color=SNM)) +
  stat_ecdf(aes(x=2-rep.fitted, color=SNM)) +
  scale_color_paletteer_d("nbapalettes::sixers_retro") +
  theme_half_open() +
  theme(legend.justification = c("right", "bottom"),) +
  xlab("Replication Timing") +
  ylab("Density")

p
```


```{r}
# inset <- ggplot(mpg, aes(drv)) + 
#   geom_bar(fill = "skyblue2", alpha = 0.7) + 
#   scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
#   theme_minimal_hgrid(11)

ggdraw(p + theme(legend.justification = c("right", "bottom"),)) +
  draw_plot(rep.plot, .1, .5, .5, .5) #+
  # draw_plot_label(
  #   c("A", "B"),
  #   c(0, 0.45),
  #   c(1, 1),
  #   size = 12
  # )

```


```{r}
rep.cuts = cut(2-d$rep.fitted, breaks=seq(0,1,0.2))
other.df <- data.frame(rep.cuts = rep.cuts, muts = d$SNM, 
                     bases=case_when(d$base.type == "diNT" ~ "CpG",
                                     d$base.type == "AT" ~ "A/T",
                                     d$base.type == "CG" ~ "C/G"), 
                     protein=case_when(d$protein ~ "Gene",
                                       !d$protein ~ "Non-Gene")
                     ) %>% group_by(rep.cuts, protein) %>% summarize(n=n(), snms=sum(muts), fraction=sum(muts)/n())

other.df$group <- as.numeric(str_extract(as.character(other.df$rep.cuts), "\\((.*),.*]", 1))
other.df

other.df[,c("est", "lower", "upper")] <- t(apply(other.df,1,function(r){BinomCI(as.numeric(r['snms']), as.numeric(r['n']), method="wilson")}))
other.df
other.df$bases = "All"

other.df %>% ggplot() +
  geom_line(aes(x=group, y=fraction/624598)) +
  # geom_point
  geom_errorbar(aes(x = group, ymin = lower/624598, ymax = upper/624598), width=0.03) +
  geom_point(aes(x=group, y=fraction/624598)) + 
  geom_smooth(aes(x=group, y=fraction/624598), method="lm", se=FALSE, lty="11") +
  scale_color_paletteer_d("nbapalettes::warriors_city", guide = guide_legend(title="Base Type")) +
  xlab("Replication Timing") +
  ylab("Mutation Rate") +
  facet_wrap(vars(protein)) +
  theme_bw(base_size=12) +
  ylim(0, 0.75e-09) 
```

```{r}
rep.cuts = cut(2-d$rep.fitted, breaks=seq(0,1,0.2))
rep.df <- data.frame(rep.cuts = rep.cuts, muts = d$SNM, 
                     bases=case_when(d$base.type == "diNT" ~ "CpG",
                                     d$base.type == "AT" ~ "A/T",
                                     d$base.type == "CG" ~ "C/G"),
                     protein=case_when(d$protein ~ "Gene",
                                       !d$protein ~ "Non-Gene")
                     ) %>% group_by(rep.cuts, bases, protein) %>% summarize(n=n(), snms=sum(muts), fraction=sum(muts)/n()) 

rep.df$group <- as.numeric(str_extract(as.character(rep.df$rep.cuts), "\\((.*),.*]", 1))
rep.df

rep.df[,c("est", "lower", "upper")] <- t(apply(rep.df,1,function(r){BinomCI(as.numeric(r['snms']), as.numeric(r['n']), method="wilson")}))
rep.df

rep.df %>% ggplot() +
  # geom_line(aes(x=group, y=fraction/624598, color=bases)) +
  # geom_point
  geom_errorbar(aes(x = group, ymin = lower/624598, ymax = upper/624598, color=as.factor(bases)), width=0.03) +
  geom_point(aes(x=group, y=fraction/624598, color=as.factor(bases))) + 
  geom_smooth(aes(x=group, y=fraction/624598, color=as.factor(bases)), method="lm", se=FALSE, lty="11") +
  scale_color_paletteer_d("nbapalettes::knicks_city", guide = guide_legend(title="Base Type")) +
  xlab("Replication Timing") +
  ylab("Mutation Rate") +
  facet_wrap(vars(protein)) +
  theme_bw(base_size=12) +
  coord_cartesian(ylim=c(0, 2.5e-09)) +
  # geom_point
  geom_errorbar(data = other.df, aes(x = group, ymin = lower/624598, ymax = upper/624598, color=as.factor(bases)), width=0.03) +
  geom_point(data = other.df, aes(x=group, y=fraction/624598, color=as.factor(bases))) + 
  geom_smooth(data = other.df, aes(x=group, y=fraction/624598, color=as.factor(bases)), method="lm", se=FALSE, lty="11")
```

```{r}
print("something")
```



```{r, fig.width = 20}
len.cuts = cut(d$length, breaks=seq(0,max(d$length),max(d$length)/15))
dist.df <- data.frame(len.cuts = len.cuts, muts = d$SNM,
                     length = d$length,
                     distance = d$distance,
                     rep.timing = d$rep.fitted > 1.5,
                     bases=case_when(d$base.type == "diNT" ~ "CpG",
                                     d$base.type == "AT" ~ "A/T",
                                     d$base.type == "CG" ~ "C/G"), 
                     protein=case_when(d$protein ~ "Gene",
                                       !d$protein ~ "Non-Gene"),
                     collision=case_when(d$collision > 0 ~ "Head-On Collision Possible",
                                       !(d$collision > 0) ~ "Head-On Collision Not Possible")
                     ) %>% 
  filter(muts) %>%
  group_by(len.cuts, protein) %>% 
  summarize(n=n(), 
            est.dist=MeanCI(distance, method="boot")["mean"],
            lower.dist=MeanCI(distance, method="boot")["lwr.ci"],
            upper.dist=MeanCI(distance, method="boot")["upr.ci"],
            est.len=MeanCI(length/2, method="boot")["mean"],
            lower.len=MeanCI(length/2, method="boot")["lwr.ci"],
            upper.len=MeanCI(length/2, method="boot")["upr.ci"])

dist.df$group <- as.numeric(str_extract(as.character(dist.df$len.cuts), "\\(.*,(.*)]", 1))
dist.df

dist.df %>% 
  ggplot() +
  geom_errorbar(aes(x = group, ymin = lower.dist, ymax = upper.dist, color=as.factor("Distance")), width=500) +
  geom_errorbar(aes(x = group, ymin = lower.len, ymax = upper.len, color=as.factor("Length")), width=500) +
  geom_point(aes(x=group, y=est.len, color=as.factor("mid"))) + 
  scale_color_paletteer_d("nbapalettes::warriors_city") +
  theme_bw(base_size=25) +
  facet_grid(vars(protein)) +
  coord_cartesian(ylim=c(0, 10000))


dist.df %>% ggplot() +
  geom_errorbar(aes(x = group, ymin = lower/624598, ymax = upper/624598), width=0.03) +
  geom_point(aes(x=group, y=fraction/624598, color=bases)) + 
  # geom_smooth(aes(x=group, y=fraction/624598, color=bases), method="lm", se=FALSE, lty="11") +
  scale_color_paletteer_d("nbapalettes::warriors_city", guide = guide_legend(title="Base Type")) +
  xlab("Distance from Start Site") +
  ylab("Mutations") +
  # facet_grid(vars(protein), vars(collision)) +
  theme_bw(base_size=12) +
  coord_cartesian(ylim=c(0, 7e-09))
```



```{r}
library(vioplot)

with(d , vioplot( 
  rep.fitted[SNM] , rep.fitted[!SNM],  
  col=rgb(0.1,0.4,0.7,0.7) , names=c("SNM","Non SNM") 
))
```



```{r}
d[seq(1, nrow(d), length.out = 10000), ] %>%
  ggplot() +
  geom_point(aes(x = rep.fitted, y = predicted, color = interaction(base.type, protein, sep = ":")))
```

```{r}
d %>%
  ggplot() +
  stat_ecdf(aes(x = 2 - rep.fitted, color = interaction(base.type, SNM, sep = ":")),
    geom = "step"
  ) +
  # stat_ecdf(data=d %>% filter(!SNM), aes(color=is.na(length), x=2-rep.fitted), geom = "step",
  #           color="blue") +
  stat_ecdf(data = data.frame(random = seq(0, 1, 0.002)), aes(x = random), color = "grey", geom = "step")
```


```{r}
dummy_df %>% ggplot() +
  geom_contour_filled(aes(x = 2 - rep.fitted, y = distance, z = -predicted),
    binwidth = 0.00001,
    show.legend = FALSE
  ) +
  geom_point(data = d[d$SNM, ], aes(x = 2 - rep.fitted, y = distance), color = "red", alpha = 0.2) +
  xlab("Replication Timing") +
  ylab("Distance from Start Site") +
  theme(text = element_text(size = 18)) +
  ggtitle("Predicted Mutation Probability")
```

```{r}
dummy_df %>% ggplot() +
  geom_contour_filled(aes(x = length, y = distance, z = -predicted),
    binwidth = 0.00001,
    show.legend = FALSE
  ) +
  geom_point(data = d[d$SNM, ], aes(x = length, y = distance), color = "red", alpha = 0.1) +
  # xlab("Replication Timing") +
  # ylab("Distance from Start Site") +
  theme(text = element_text(size = 18)) +
  ggtitle("Predicted Mutation Probability")
```

```{r}
d %>%
  ggplot() +
  stat_ecdf(aes(x = 2 - rep.fitted, color = SNM), geom = "step") +
  stat_ecdf(data = data.frame(random = seq(0, 1, 0.002)), aes(x = random), color = "grey", geom = "step") +
  scale_color_viridis_d() +
  xlab("Replication Timing") +
  ylab("Cumulative density") +
  theme(text = element_text(size = 18))
```

```{r}
d %>%
  ggplot() +
  geom_density(aes(x = distance / length, color = SNM))
```

```{r}
dummy_df %>% ggplot() +
  geom_point(aes(x = distance, y = predicted))
```

```{r}
dummy_df %>% ggplot() +
  geom_point(aes(x = distance, y = predicted, color = interaction(protein, collision, sep = ":"))) +
  scale_color_viridis_d()
```

```{r}
n <- 20
dummy_df <- expand.grid(
  base.type = c("CG", "diNT", "AT"),
  rep.fitted = mean(d$rep.fitted),
  protein = c(TRUE, FALSE),
  distance = seq(min(d$distance), max(d$distance), length.out = n),
  collision = c(1, 0),
  transcripts = mean(d$transcripts),
  length = mean(d$length)
)

dummy_df$predicted <- predict(mod, newdata = dummy_df, type = "response")
```


```{r}
dummy_df %>% ggplot() +
  geom_line(aes(x = distance, y = predicted, color = base.type)) +
  facet_grid(rows=vars(protein), cols=vars(collision)) +
  scale_color_viridis_d()
```



```{r}
dummy_df %>% ggplot() +
  geom_line(aes(x = distance, y = predicted, color = interaction(protein, collision, sep = ":"))) +
  geom_density(data = d[d$SNM & d$protein & d$collision, ], aes(x = distance), color = "red", alpha = 0.1) +
  geom_density(data = d[d$SNM & !d$protein & d$collision, ], aes(x = distance), color = "orange", alpha = 0.1) +
  geom_density(data = d[d$SNM & d$protein & !d$collision, ], aes(x = distance), color = "pink", alpha = 0.1) +
  geom_density(data = d[d$SNM & !d$protein & !d$collision, ], aes(x = distance), color = "brown", alpha = 0.1) +
  scale_color_viridis_d()
```

```{r}
dummy_df %>% ggplot() +
  geom_line(aes(x = length, y = predicted, color = interaction(protein, collision, sep = ":"))) +
  scale_color_viridis_d()
```


```{r}
dummy_df %>% ggplot() +
  geom_line(aes(x = transcripts, y = predicted, color = interaction(protein, collision, sep = ":"))) +
  scale_color_viridis_d()
```


```{r}
dummy_df %>% ggplot() +
  geom_line(aes(x = rep.fitted, y = predicted, color = interaction(protein, collision, sep = ":"))) +
  scale_color_viridis_d()
```


```{r}
t.test(d[d$SNM, ]$distance / d[d$SNM, ]$length, mu = 0.5)
```


```{r}
dist.df <- d[d$SNM, c("SNM", "distance", "length", "predicted")]
dist.df <- dist.df %>% arrange("length")
dist.df %>%
  group_by(L = cut(length, breaks = 15), D = cut(distance, breaks = 1000)) %>%
  summarize(observed = sum(SNM)) %>%
  mutate(
    distance_group = as.numeric(str_extract(D, "\\((.*),.*]", 1)),
    length_group = as.numeric(str_extract(L, "\\(.*,(.*)]", 1))
  ) %>%
  ggplot() +
  geom_point(aes(x = distance_group, y = observed)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  facet_grid(cols = vars(length_group), scales = "free") #+
# xlim(0, length_group)



# q <- t %>%
#   filter(!is.na(d$distance)) %>%
#   group_by(G = cut(predict_all, breaks=unique(quantile(predict_all, probs=seq(0, 1, by=0.01))))) %>%
#   summarize(predicted = mean(predict_all), observed = mean(SNM), count=n())
```


```{r}
dist.df <- d[d$SNM, c("SNM", "distance", "length", "predicted")]

breaks <- as.numeric(str_extract(unique(cut(dist.df$length, breaks = 10)), "\\(.*,(.*)]", 1))

dist.df %>%
  group_by(L = cut(length, breaks = 10), SNM) %>%
  summarize(mean(distance), sd(distance), n = n(), mean(length), sd(length)) %>%
  pivot_wider(names_from = SNM, values_from = c(
    "mean(distance)", "sd(distance)", "n",
    "mean(length)", "sd(length)"
  )) %>%
  mutate(length_group = as.numeric(str_extract(L, "\\(.*,(.*)]", 1))) %>%
  mutate(
    moe = qnorm(1 - 0.05 / 2) * `sd(distance)_TRUE` / sqrt(n_TRUE),
    moe_l = qnorm(1 - 0.05 / 2) * `sd(length)_TRUE` / sqrt(n_TRUE)
  ) %>%
  mutate(
    upper = `mean(distance)_TRUE` + moe,
    lower = `mean(distance)_TRUE` - moe,
    u = `mean(length)_TRUE` + moe_l,
    l = `mean(length)_TRUE` - moe_l
  ) %>%
  ggplot() +
  geom_errorbar(aes(x = length_group, y = `mean(distance)_TRUE`, ymin = lower, ymax = upper), color = "red") +
  geom_errorbar(aes(x = length_group, y = `mean(length)_TRUE`, ymin = l, ymax = u)) +
  geom_step(aes(x = length_group, y = length_group), lines = "horizontal", direction = "mid", color = "grey") +
  # geom_step(aes(x = length_group, y = length_group/2), lines = "horizontal", direction = "mid", color = "grey") +
  geom_step(aes(x = length_group, y = `mean(length)_TRUE`/2), lines = "horizontal", direction = "mid", color = "grey") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  xlab("Length group") +
  scale_x_continuous(
    labels = breaks,
    breaks = breaks
  ) +
  ylab("Distance from Start Site")
```


```{r}
d %>% ggplot() +
  geom_hex(aes(x = transcripts, y = rep.fitted), bins = 30)
d %>% ggplot() +
  geom_hex(aes(x = transcripts, y = rep.fitted, fill = SNM), bins = 30)
```


```{r}
# library(car)
# avPlots(mod, layout=c(1,3))
```

```{r}
# coeffs$PVE <- c(0.8669414,2,3,4,5,6,7,8,9,10,11,12,13,14)
coeffs <- data.frame(
  Predictors = c(
    "base.type",
    "rep",
    "protein",
    "distance",
    "collision",
    "distance^2",
    "base:collision",
    "transcripts:length",
    "protein:distance",
    "base:distance",
    "distance:collision"
  ),
  PVE = c(
    0.8712471,
    0.01976554,
    0.0678921,
    0.05402601,
    0.0207636,
    0.008077868,
    0.005247945,
    0.00368679,
    0.02515068,
    0.01279865,
    0.003516638
  )
)

coeffs %>%
  ggplot() +
  geom_col(aes(y = PVE * 100, x = fct_reorder(Predictors, PVE, .desc = TRUE))) +
  theme(
    text = element_text(size = 18),
    axis.text.x = element_text(angle = 90, hjust = 1)
  ) +
  ylab("PVE") +
  xlab("Predictors")
```



```{r}
df <- read.csv("/outs/by_feature.csv")
head(df)
```



```{r}
snms <- d[d$SNM,]
```

```{r}
counts <- c()
for (i in 1:nrow(df)){
  lower <- df[i,"start"]
  upper <- df[i,"end"]
  # print(df[i,])
  # print(snms[(snms$pos>=lower) & (snms$pos<=upper),])
  counts <- c(counts, nrow(snms[(snms$pos>=lower) & (snms$pos<=upper),]))
}
# counts
df$SNM <- counts
```


```{r}
dim(df)
df <- df %>% drop_na()
df <- df %>% filter(!str_ends(type, "opp")) %>% filter(!str_ends(type, "same"))
dim(df)
df %>% head()
```

```{r}
df %>% ggplot() +
  geom_histogram(aes(x=SNM))
```

```{r}
df %>% ggplot() +
  geom_histogram(aes(x=SNM/length))
```


```{r}
df %>% 
  ggplot() +
  geom_point(aes(x=length, y=SNM+1)) +
  scale_x_log10() +
  scale_y_log10()
```


```{r}
var(df$SNM); mean(df$SNM)
```

```{r}
df %>% filter(SNM > 30)
```


```{r}
model1 <- glm(SNM ~ log(length), data=df, family=poisson(link=log))
summary(model1)
```


```{r}
model2 <- glm.nb(SNM ~ log(length), data=df)
summary(model2)
```

Over-dispersion param is not 0

```{r}
model3 <- glm(SNM ~ log(length), data=df, family=quasipoisson(link=log))
summary(model3)
```

```{r}
model4 <- glm(SNM ~ base.type + log(rep.fitted) + offset(log(length)), data=df, family=quasipoisson(link=log))
summary(model4)
anova(model4)
```
mod.NB.GC <- glm.nb(Non ~ offset(I(1*log(L))) + log(gc), data = data) # model WITH gc included
mod.NB.0 <- glm.nb(Non ~ offset(I(1*log(L))), data = data) # model WITHout gc included

require(lmtest)
lrtest(mod.NB.GC, mod.NB.0)

```{r}
library(lmtest)
model5 <- glm.nb(SNM ~ log(base.type) + offset(log(length)), data=df)
summary(model5)

model6 <- glm.nb(SNM ~ offset(log(length)), data=df)
summary(model6)
lrtest(model6, model5)
```

```{r}
AIC(model5)
AIC(model6)
```


```{r}
model7 <- glm.nb(SNM ~ log(base.type) + log(reads)*log(length) + offset(log(length)), data=df)
summary(model7)
anova(model7)
lrtest(model7, model5)
```


```{r}
model8 <- glm.nb(SNM ~ log(base.types) + log(reads) + offset(log(length)), data=df)
summary(model8)
anova(model8)
lrtest(model8, model5)
```


```{r}
pca <- as.data.frame(prcomp(~ (reads + length + base.type + rep.fitted + collision)^2, data=df, scale.=TRUE)$x)

model9 <- glm.nb(df$SNM ~ (.), data=pca)
summary(model9)
```
```{r}
full <- glm.nb(SNM ~ collision + offset(log(length)), data=df)
null <- glm.nb(SNM ~ offset(log(length)), data=df)

lrtest(null, full)
```

reads + length + base.type + rep.fitted + collision + protein

```{r}
model10 <- glm.nb(SNM ~ base.type + rep.fitted + length*reads + collision, data=df)
summary(model10)

str(df)
```




```{r}
model1 <- glm(SNM ~ length + reads + length*reads + offset(log(df$length)), data=df, family="poisson")
summary(model1)
```

```{r}
model2 <- glm(SNM ~ length + reads + length*reads + offset(log(df$length)), data=df, family="quasipoisson")
summary(model2)
```


```{r}
library(MASS)
```

```{r}
model3 <- glm.nb(SNM ~ length + reads + length*reads + offset(log(df$length)), data=df)
summary(model3)
```

```{r}
pchisq(2 * (logLik(model3) - logLik(model1)), df = 1, lower.tail = FALSE)
```



```{r}
model4 <- glm(SNM ~ length:reads + offset(log(df$length)), data=df, family="quasipoisson")
summary(model4)
```




```{r}
model5 <- glm(SNM ~ length + reads + length:reads + offset(log(df$length)), data=df, family="quasipoisson")
summary(model5)
```

```{r}
model6 <- glm(SNM ~ base.type + rep.fitted + protein + collision + reads + length + base.type:protein + base.type:reads + base.type:length + rep.fitted:reads + rep.fitted:length + protein:reads + protein:length + collision:length + reads:length, data=df, family="quasipoisson")
summary(model6)
```

```{r}
model7 <- glm(SNM ~ base.type + rep.fitted +  collision + reads + length + base.type:reads + base.type:length + rep.fitted:length + offset(log(df$length)), data=df, family="quasipoisson")
summary(model7)
```

```{r}
model8 <- glm(SNM ~ base.type + rep.fitted + offset(log(length)), data=df, family="quasipoisson")
summary(model8)
```

```{r}
min(df$rep.fitted)
```


```{r}
# df$predicted <- predict(model6, newdata=df, type="response")

max(df$base.type)
df %>% filter((SNM/length)>0.03)

df %>% ggplot() +
  geom_point(aes(x=base.type, y=SNM/length))

df %>% ggplot() +
  geom_point(aes(x=base.type, y=SNM))

df %>% ggplot() +
  geom_point(aes(x=length, y=SNM/length))

df %>% ggplot() +
  geom_point(aes(x=length, y=SNM))

df %>% ggplot() +
  geom_point(aes(x=log(reads), y=SNM/length))

df %>% ggplot() +
  geom_point(aes(x=log(reads), y=SNM))

df %>% ggplot() +
  geom_point(aes(x=rep.fitted, y=SNM/length))

df %>% ggplot() +
  geom_point(aes(x=rep.fitted, y=SNM))

df %>% ggplot() +
  geom_point(aes(x=protein, y=SNM/length), position="jitter")

df %>% ggplot() +
  geom_point(aes(x=protein, y=SNM), position="jitter")
```

```{r}
cor(df$SNM, df$rep.fitted)
```


